---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: ../style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 200,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)
```

# __Functional CpG annotation__
***

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/util.R", chdir = TRUE)

library(data.table)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(scales)
library(GGally)
library(factoextra)
library(viridis)
library(knitr)

library(splitstackshape)
library(reshape2)
library(IRanges)
library(ggplot2)
library(biomaRt)

require(foreign)
library(parallel)
library(foreach)
library(doParallel)

library(gUtils)

out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/region_wise_independent_snps/"

# meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_dex_with_delta_fdr_005.csv")
# meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_veh_fdr_005.csv")
# meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_result_delta_fdr_005.csv")

# col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

# ind.meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
# ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
# ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r include = F, eval = T}
# code: ../code/methylation/02_dma/01_dmp/03_get_cpg_gene_anno_dist
cpg.closest.genes.dist.fn <-"~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_cpgs_annotated_closest_genes_distances.csv"
cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)

delta.meqtl.cpg.anno.df               <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_delta.csv"), sep = "\t")
delta.meqtl.cpg.closest.genes.dist.df <- cpg.closest.genes.dist.df[PROBE_ID %in% delta.meqtl.cpg.anno.df$Name]

veh.meqtl.cpg.anno.df                 <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_veh.csv"), sep = "\t")
veh.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% veh.meqtl.cpg.anno.df$Name]

dex.meqtl.cpg.anno.df                 <- fread(paste0(out.dir.pre, "meqtls_cpg_annotated_dex.csv"), sep = "\t")
dex.meqtl.cpg.closest.genes.dist.df   <- cpg.closest.genes.dist.df[PROBE_ID %in% dex.meqtl.cpg.anno.df$Name]
# delta.meqtl.cpg.closest.genes.dist.fn <-"~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis/delta/annotation/me-qtl_delta_cpgs_annotated_closest_genes_distances.csv"
# delta.meqtl.cpg.closest.genes.dist.df <- fread(cpg.closest.genes.dist.fn)
```

## Illumina {.tabset .tabset-fade .tabset-pills}

### Example of annotated CpGs {.tabset .tabset-fade .tabset-pills}

#### Full anno

```{r include = T, eval = T}
delta.meqtl.cpg.anno.df$Relation_to_Island <- factor(delta.meqtl.cpg.anno.df$Relation_to_Island, 
                                                     levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
delta.meqtl.cpg.anno.df[["Model"]] <- as.factor("Delta")

veh.meqtl.cpg.anno.df$Relation_to_Island <- factor(veh.meqtl.cpg.anno.df$Relation_to_Island,
                                                   levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
veh.meqtl.cpg.anno.df[["Model"]] <- as.factor("Baseline")

dex.meqtl.cpg.anno.df$Relation_to_Island <- factor(dex.meqtl.cpg.anno.df$Relation_to_Island,
                                                   levels = c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea"))
dex.meqtl.cpg.anno.df[["Model"]] <- as.factor("Dex")

meqtl.cpg.anno.df <- rbind(delta.meqtl.cpg.anno.df, veh.meqtl.cpg.anno.df, dex.meqtl.cpg.anno.df)

head(meqtl.cpg.anno.df, 100)
```

#### Closest genes

```{r include = T, eval = T}
head(cpg.closest.genes.dist.df, 100)
```

### Chromosomes

```{r out.width = "90%", fig.cap = "CpGs distribution across chromosomes", include = T, eval = T}
library(GenomicFeatures)
chr.length.tbl <- getChromInfoFromUCSC("hg19")[, c(1, 2)] %>% setDT()
chr.length.tbl <- chr.length.tbl[chrom %in% paste0("chr", 1:22)] #, "X", "Y"))]
colnames(chr.length.tbl) <- c("chr", "chr_size")
chr.length.tbl$chr <- as.factor(chr.length.tbl$chr)

chr.order <- paste("chr", c(1:22), sep = "")
meth      <- delta.meqtl.cpg.anno.df
meth$chr  <- factor(meth$chr,levels = chr.order)
meth      <- left_join(meth, chr.length.tbl)

library(janitor)
chr.cpgs.cnt <- tabyl(meth$chr, sort = T)
colnames(chr.cpgs.cnt) <- c("chr", "cnt_cpg", "freq_cpg")

chr.cpgs.cnt <- left_join(chr.cpgs.cnt, chr.length.tbl, by = "chr")
chr.cpgs.cnt$chr  <- factor(chr.cpgs.cnt$chr,levels = chr.order)
chr.cpgs.cnt[["freq_chr"]] <- chr.cpgs.cnt$chr_size / sum(chr.cpgs.cnt$chr_size)
chr.cpgs.cnt[["cpg_chr_cnt"]] <- chr.cpgs.cnt$cnt_cpg / chr.cpgs.cnt$freq_chr
chr.cpgs.cnt[["cpg_chr_freq"]] <- chr.cpgs.cnt$cpg_chr_cnt / sum(chr.cpgs.cnt$cpg_chr_cnt)

# Relative freq = (Number of CpGs on the chromosome (i) * (length of chromosomes{1:22} / length of the chromosome(i)) ) in %

ggplot(chr.cpgs.cnt, aes(x = chr, y = cpg_chr_freq, fill = chr)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  geom_text(aes(label = scales::percent(cpg_chr_freq, accuracy = 0.1), y = cpg_chr_freq), 
            stat = "identity", vjust = -.5, size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(title = "Delta meQTLs CpGs distribution across the chromosomes wieghted by the chromosomes' length",
       x = "Chromosome",
       y = "Relative frequency") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 8),
         axis.title = element_text(size = 8),
         axis.text.x = element_text(angle = 20, hjust = 0.5),
         legend.position = "none")
```

### Genomic regions

```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

```{r}

delta <- setdiff(setdiff(meqtls.cpgs$delta, meqtls.cpgs$dex), meqtls.cpgs$veh)
veh <- setdiff(setdiff(meqtls.cpgs$veh, meqtls.cpgs$dex), meqtls.cpgs$delta)
dex <- setdiff(setdiff(meqtls.cpgs$dex, meqtls.cpgs$veh), meqtls.cpgs$delta)

meqtl.unique.cpg.anno.df <- rbind(  
  delta.meqtl.cpg.anno.df[Name %in% delta,],
  veh.meqtl.cpg.anno.df[Name %in% veh,],
  dex.meqtl.cpg.anno.df[Name %in% dex,]
)
```


```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.unique.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "Unique meCpGs") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

```{r}

intersected <- intersect(intersect(meqtls.cpgs$delta, meqtls.cpgs$dex), meqtls.cpgs$veh)

meqtl.intersected.cpg.anno.df <- rbind(  
  delta.meqtl.cpg.anno.df[Name %in% intersected,],
  veh.meqtl.cpg.anno.df[Name %in% intersected,],
  dex.meqtl.cpg.anno.df[Name %in% intersected,]
)
```

```{r out.width = "90%", fig.cap = "Distribution of locations of the CpGs in different genomic regions based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c( "#CC6677", "#DDCC77", "#117733", "#AA4499", "#44AA99", "#888888")

ggplot(meqtl.intersected.cpg.anno.df, aes(x = Model, fill = Relation_to_Island)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model, # scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1), 
                y = (..count..)/sum(..count..)), 
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "Intersected meCpGs") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 12),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```

### Gene regions

```{r include = F, eval = T}
delta.meqtl.cpg.closest.genes.dist.df$GeneGroup <- as.factor(delta.meqtl.cpg.closest.genes.dist.df$GeneGroup)

delta.meqtl.cpg.closest.genes.dist.df[["Model"]] <- as.factor("Delta")
veh.meqtl.cpg.closest.genes.dist.df[["Model"]]   <- as.factor("Baseline")
dex.meqtl.cpg.closest.genes.dist.df[["Model"]]   <- as.factor("Dex")

meqtl.cpg.closest.genes.dist.df <- rbind(delta.meqtl.cpg.closest.genes.dist.df,
                                         veh.meqtl.cpg.closest.genes.dist.df,
                                         dex.meqtl.cpg.closest.genes.dist.df)

meqtl.cpg.closest.genes.dist.df[is.na(meqtl.cpg.closest.genes.dist.df$GeneGroup), GeneGroup := "#NA"]
```

```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the CpGs based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.cpg.closest.genes.dist.df, aes(x = Model, fill = GeneGroup)) + 
  geom_bar( position = "fill") + 
  geom_text(aes(by = Model), # label = scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1)),
            stat = "prop", 
            position = position_fill(vjust = .5), size = 3) +
 # scale_y_continuous(labels = scales::percent) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "Annotation to the closest gene based on IlluminaHumanMethylationEPICanno.ilm10b4.hg19 package") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Type", values = cbPalette)
```

```{r include = F, eval = T}
# Prepare methyl ranges

cpg.coord.df    <- delta.meqtl.cpg.anno.df[, .(Name, chr, pos)]# %>% dplyr:: mutate(chr = sub("chr", "", chr))
cpg.coord.range <- makeGRangesFromDataFrame(cpg.coord.df, 
                                            start.field = "pos", 
                                            end.field = "pos", 
                                            seqnames.field = c("chr"))
names(cpg.coord.range) <- cpg.coord.df$Name
```

```{r include = F, eval = F}
library(annotatr)
library(AnnotationHub)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

annots <- c('hg19_cpgs', 'hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries')
gene.annots <- c('hg19_basicgenes') #, 'hg19_genes_intergenic', 'hg19_genes_intronexonboundaries')
    
annotations <- build_annotations(
  genome = "hg19",
  annotations = gene.annots
)
```

```{r include = F, eval = F}
delta.meqtls.cpg.annotated <- annotate_regions(
    regions = cpg.coord.range,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE)

# print(delta.meqtls.cpg.annotated)
# plot_annotation(annotated_regions = delta.meqtls.cpg.annotated)

delta.meqtls.cpg.annotated.df <- data.frame(delta.meqtls.cpg.annotated) %>% setDT()

delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_1to5kb", annot.type := "1-5kb"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_3UTRs", annot.type := "3UTRs"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_5UTRs", annot.type := "5UTRs"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_exons", annot.type := "exons"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_introns", annot.type := "introns"]
delta.meqtls.cpg.annotated.df[annot.type == "hg19_genes_promoters", annot.type := "promoters"]
# head(delta.meqtls.cpg.annotated.df)

delta.meqtls.cpg.annotated.df <- delta.meqtls.cpg.annotated.df[, .(seqnames, start, end, annot.type)] %>% unique()

delta.meqtls.cpg.annotated.df$annot.type <- factor(delta.meqtls.cpg.annotated.df$annot.type, 
                                                              levels = c("1-5kb", "promoters", "5UTRs", "introns", "exons", "3UTRs"))
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = F}
ggplot(delta.meqtls.cpg.annotated.df, aes(x = annot.type)) + 
  geom_bar(aes(y = (..count..)/sum(..count..), fill = annot.type), position = position_dodge()) + 
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Gene Region",
       y = "Percentage of CpG", 
       title = "From UCSC Genome Browser") +
  theme(legend.title = element_blank(), 
        legend.position = "none", #c(.1,.8),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
   scale_fill_manual(values = cbPalette)
```

Annotation with Chipseeker

```{r include = F, eval = T}
delta.meqtl.snp.anno.rds              <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_delta.rds"))
veh.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_veh.rds"))
dex.meqtl.snp.anno.rds                <- readRDS(paste0(out.dir.pre, "meqtls_cpg_annotated_withChIPseeker_dex.rds"))
```

```{r include = F, eval = T}
delta.meqtl.snp.anno.stat.df <- delta.meqtl.snp.anno.rds@annoStat
veh.meqtl.snp.anno.stat.df   <- veh.meqtl.snp.anno.rds@annoStat
dex.meqtl.snp.anno.stat.df   <- dex.meqtl.snp.anno.rds@annoStat

delta.meqtl.snp.anno.stat.df[["Model"]] <- as.factor("Delta")
veh.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Baseline")
dex.meqtl.snp.anno.stat.df[["Model"]]   <- as.factor("Dex")

meqtl.snp.anno.stat.df <- rbind(delta.meqtl.snp.anno.stat.df,
                                veh.meqtl.snp.anno.stat.df,
                                dex.meqtl.snp.anno.stat.df)

meqtl.snp.anno.stat.df$Feature   <- factor(meqtl.snp.anno.stat.df$Feature)
meqtl.snp.anno.stat.df$Frequency <- signif(meqtl.snp.anno.stat.df$Frequency, 3)
```

```{r out.width = "90%", fig.cap = "Distribution of gene-centric locations of the CpGs based on Illumina EPIC annotation", include = T, eval = T}
cbPalette <- c("#88CCEE", "#CC6677", "#DDCC77", "#117733", "#332288", "#AA4499", "#44AA99", "#999933", "#882255", "#661100", "#6699CC", "#888888")

ggplot(meqtl.snp.anno.stat.df, aes(x = Model, y = Frequency, fill = Feature)) + 
  geom_bar( position = "fill", stat = "identity") + 
  geom_text(data = subset(meqtl.snp.anno.stat.df, Frequency > 2),
            aes(by = Model, label = Frequency), 
            position = position_fill(vjust = .5), size = 3) +
  labs(x = "",
       y = "Percentage of CpGs", 
       title = "meSNP annotation from UCSC for the hg19 genome build using TxDb.Hsapiens.UCSC.hg19.knownGene and ChIPseeker Bioconductor R packages") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 0, hjust = 0.5)) +
   scale_fill_manual("Feature", values = cbPalette)
```


```{r include = F, eval = T}
# chromhmm.all.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_all_states.Rds")
# colnames(values(chromhmm.all.states))[1] <- "type"

chromhmm.states     <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromhmm_states.Rds") %>% setDT()
chromhmm.states[COLOR.NAME == "Silver", COLOR.NAME := "Grey"] 
chromhmm.states[COLOR.NAME == "White", COLOR.NAME := "Snow"] 

cbPalette <- chromhmm.states$COLOR.NAME

chromhmm.epigenomes <- read.csv("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv", sep = "\t")
```

# __Functional CpG enrichment__
***

## Chromatin States
<!-- {.tabset .tabset-fade .tabset-pills} -->

```{r eval = T, include = F}
chromhmm.blood.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_blood_states.Rds") 
chromhmm.brain.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_brain_states.Rds") 

states.lst <- elementMetadata(chromhmm.blood.states)[, "type"] %>% unique() %>% sort()

# length(chromhmm.blood.states)
# length(chromhmm.brain.states)
```

```{r eval = T, include = F}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/05_me-qtl_enrichment/region_wise_independent_snps/"
chromhmm.blood.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_cpgs_chromHMM_blood_enrichment_perm_100.csv"))
chromhmm.brain.enrich.perm.rslt <- read.csv2(paste0(out.dir.pre, "meqtl_cpgs_chromHMM_brain_enrichment_perm_100.csv"))

chromhmm.blood.enrich.perm.rslt$state <- factor(chromhmm.blood.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
levels( chromhmm.blood.enrich.perm.rslt$state) <- chromhmm.states$DESCRIPTION

chromhmm.brain.enrich.perm.rslt$state <- factor(chromhmm.brain.enrich.perm.rslt$state, levels = chromhmm.states$STATE)
levels(chromhmm.brain.enrich.perm.rslt$state) <- chromhmm.states$DESCRIPTION
```

### Blood

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permutation Fisher-tests empirical P <= 0.05. ", eval = T, include = T}

chromhmm.blood.enrich.perm.rslt["is_sign"] <- ifelse(chromhmm.blood.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(chromhmm.blood.enrich.perm.rslt, aes(x = state, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```


```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meCpGs. The Y-axis denotes the proportion of overlaps between meCpGs and blood cell type states.", eval = T, include = T}

chromhmm.blood.enrich.perm.rslt["prop"] <-  signif(chromhmm.blood.enrich.perm.rslt$n_overlap / delta.meqtl.snp.anno.rds@peakNum * 100, 2)

ggplot(chromhmm.blood.enrich.perm.rslt, aes(x = state, y = prop)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  # geom_text(aes(label = paste0(n_snps_overlap, " \n(", prop, "%)")), position = position_dodge(width = 0.9), vjust = -0.25, cex = 2) + 
  geom_text(aes(label = n_overlap), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  labs(x = "",
       y = "Overlaps, %", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
   scale_fill_manual("", values = c( "#009E73")) 
```

### Brain

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meCpGs.  The Y-axis denotes the fold enrichment/depletion as compared to baseline meCpGs. Green bars indicate significant enrichment/depletion, grey bars non-significant differences based on permuation Fisher-tests emperical P <= 0.05.", eval = T, include = T}

chromhmm.brain.enrich.perm.rslt["is_sign"] <- ifelse(chromhmm.brain.enrich.perm.rslt$p_val_emp < 0.05, "Significant", "Non-significant")

ggplot(chromhmm.brain.enrich.perm.rslt, aes(x = state, y = odds.ratio, fill = is_sign)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
  geom_hline(yintercept = 1, color = "red") + 
  labs(x = "",
       y = "Odds ratio", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
   scale_fill_manual("", values = c( "#999999", "#009E73")) 
```

```{r out.width = "90%", fig.cap = "Histone mark enrichment for the delta meCpGs. The Y-axis denotes the proportion of overlaps between meCpGs and brain cell type states.", eval = T, include = T}

chromhmm.brain.enrich.perm.rslt["prop"] <- signif(chromhmm.brain.enrich.perm.rslt$n_overlap / delta.meqtl.snp.anno.rds@peakNum * 100, 2)

ggplot(chromhmm.brain.enrich.perm.rslt, aes(x = state, y = prop)) + 
  geom_bar( stat = "identity", position = position_dodge()) +
# geom_text(aes(label = paste0(n_snps_overlap, " (", prop, "%)")), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  geom_text(aes(label = n_overlap), position = position_dodge(width = 0.9), vjust = -0.25, cex = 3) + 
  labs(x = "",
       y = "Overlaps, %", 
       title = "") +
  theme(legend.position = "bottom",
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1)) +
   scale_fill_manual("", values = c( "#009E73")) 
```


<!-- ## Chromatin States {.tabset .tabset-fade .tabset-pills} -->

<!-- ### Annotation table -->

<!-- ```{r include = F, eval = T} -->
<!-- # chromhmm.all.states <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromHMM_all_states.Rds") -->
<!-- # colnames(values(chromhmm.all.states))[1] <- "type" -->

<!-- chromhmm.states     <- readRDS("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/chromhmm_states.Rds") %>% setDT() -->
<!-- chromhmm.states[COLOR.NAME == "Silver", COLOR.NAME := "Grey"]  -->
<!-- chromhmm.states[COLOR.NAME == "White", COLOR.NAME := "Snow"]  -->

<!-- cbPalette <- chromhmm.states$COLOR.NAME -->

<!-- chromhmm.epigenomes <- read.csv("~/bio/code/mpip/dex-stim-human-array/data/annotation/chromHMM/epigenomes.tsv", sep = "\t") -->
<!-- ``` -->

<!-- ```{r include = T, eval = T} -->
<!-- meqtls.cpg.chromhmm.annotated <- readRDS(paste0(out.dir.pre, "meqtls_cpg_chromhmm_annotated_all_models.rds")) -->

<!-- meqtls.cpg.chromhmm.annotated.df <- data.frame(meqtls.cpg.chromhmm.annotated) %>% setDT() -->
<!-- meqtls.cpg.chromhmm.annotated.df[["CpG_ID"]] <- names(meqtls.cpg.chromhmm.annotated) -->

<!-- head(meqtls.cpg.chromhmm.annotated.df %>% dplyr::select(CpG_ID, everything()), 100)  -->
<!-- # plot_annotation(annotated_regions = delta.meqtls.cpg.chromhmm.annotated) -->
<!-- ``` -->

<!-- ```{r func-plot-chromhmm-states, include = F, eval = T} -->
<!-- GetChromHMMPlot <- function(delta.meqtls.cpg.chromhmm.annotated.df, cbPalette = NULL, plot.title = NULL){ -->

<!--   delta.meqtls.cpg.chromhmm.annotated.df$annot.type <- factor(delta.meqtls.cpg.chromhmm.annotated.df$annot.type,  -->
<!--                                                               levels = chromhmm.states$STATE) -->

<!--   ggplot(delta.meqtls.cpg.chromhmm.annotated.df, aes(x = annot.type)) +  -->
<!--     geom_bar(aes(y = (..count..)/sum(..count..), fill = annot.type), position = position_dodge()) +  -->
<!--     geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)),  -->
<!--               stat = "count", vjust = -.5,  position = position_dodge(1), size = 3) + -->
<!--     scale_y_continuous(labels = scales::percent) + -->
<!--     labs(x = "ChromHMM states", -->
<!--          y = "Percentage of CpG",  -->
<!--          title = plot.title) + -->
<!--     theme(legend.title = element_blank(),  -->
<!--           legend.position = "none", #c(.1,.8), -->
<!--           panel.grid.major = element_blank(), -->
<!--           panel.background = element_blank(), -->
<!--           plot.title = element_text(size = 8), -->
<!--           axis.title = element_text(size = 8), -->
<!--           axis.text.x = element_text(angle = 45, hjust = 0.5)) + -->
<!--     scale_fill_manual(values = cbPalette) -->
<!-- } -->

<!-- GetChromHMMPlotAll <- function(df, cbPalette = NULL, plot.title = NULL){ -->

<!--   df$annot.type <- factor(df$annot.type, levels = chromhmm.states$STATE) -->
<!--   df$Model      <- factor(df$Model, levels = c("Delta", "Baseline", "Dex")) -->

<!--   ggplot(df, aes(x = Model, fill = annot.type)) +  -->
<!--     geom_bar( position = "fill") +  -->
<!--     # coord_flip() + -->
<!--     # scale_y_reverse() + -->
<!--     geom_text(aes(by = Model), # label = scales::percent(..count../tapply(..count.., ..x.., sum), accuracy = 0.1)), -->
<!--               stat = "prop",  -->
<!--               position = position_fill(vjust = .5), size = 3) + -->
<!--     labs(x = "", -->
<!--        y = "Percentage of CpGs",  -->
<!--        title = plot.title) + -->
<!--     theme(legend.position = "right", # "bottom", -->
<!--           panel.grid.major = element_blank(), -->
<!--           panel.background = element_blank(), -->
<!--           plot.title = element_text(size = 8), -->
<!--           axis.title = element_text(size = 8)) + -->
<!--           # axis.text.x = element_text(angle = 0, hjust = 0.5)) + -->
<!--     scale_fill_manual("", values = cbPalette)  -->
<!--     # guides(fill = guide_legend(nrow = 2, byrow = TRUE)) -->
<!-- } -->
<!-- ``` -->

<!-- ```{r read-chromhmm-anno-delta, eval = T, include = F} -->
<!-- delta.meqtls.cpg.chromhmm.annotated.df <- fread(paste0(out.dir.pre, "meqtls_cpg_chromhmm_annotated_delta.csv")) -->

<!-- delta.meqtls.cpg.chromhmm.annotated.df[, Model := "Delta"] -->
<!-- ``` -->

<!-- ```{r read-chromhmm-anno-veh, eval = T, include = F} -->
<!-- veh.meqtls.cpg.chromhmm.annotated.df <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls_cpg_chromhmm_annotated_veh.csv") -->

<!-- veh.meqtls.cpg.chromhmm.annotated.df[, Model := "Baseline"] -->
<!-- ``` -->

<!-- ```{r read-chromhmm-anno-dex, eval = T, include = F} -->
<!-- dex.meqtls.cpg.chromhmm.annotated.df <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/meqtls_cpg_chromhmm_annotated_dex.csv") -->

<!-- dex.meqtls.cpg.chromhmm.annotated.df[, Model := "Dex"] -->
<!-- ``` -->

<!-- ```{r eval = T, include = F} -->
<!-- meqtls.cpg.chromhmm.annotated.plt.df <- rbind(delta.meqtls.cpg.chromhmm.annotated.df,  -->
<!--                                               veh.meqtls.cpg.chromhmm.annotated.df, -->
<!--                                               dex.meqtls.cpg.chromhmm.annotated.df) -->
<!-- ``` -->

<!-- ### All cell types -->

<!-- ```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 127 cell types", include = T, eval = T} -->

<!-- df <- meqtls.cpg.chromhmm.annotated.plt.df[, .(CpG_ID, annot.type, Model)] %>% unique() -->

<!-- plot.title <- "15 chromatin states from 127 cell types" -->

<!-- GetChromHMMPlotAll(df, cbPalette, plot.title) -->
<!-- ``` -->

<!-- ### Blood and T-cell -->

<!-- ```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 14 blood and T-cell types", include = T, eval = T} -->

<!-- chromhmm.blood.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Blood & T-cell", "EID"] -->

<!-- df <- meqtls.cpg.chromhmm.annotated.plt.df[annot.code %in% chromhmm.blood.eids, .(CpG_ID, annot.type, Model)] %>% unique() -->

<!-- plot.title <- "15 chromatin states from 14 blood and T-cell types" -->

<!-- GetChromHMMPlotAll(df, cbPalette, plot.title) -->
<!-- ``` -->

<!-- ### Brain -->

<!-- ```{r out.width = "90%", fig.cap = "The distribution of CpGs in 15 chromatin states from 10 brain types", include = T, eval = T} -->
<!-- chromhmm.brain.eids <- chromhmm.epigenomes[chromhmm.epigenomes$GROUP == "Brain", "EID"] -->

<!-- df <- meqtls.cpg.chromhmm.annotated.plt.df[annot.code %in% chromhmm.brain.eids, .(CpG_ID, annot.type, Model)] %>% unique() -->

<!-- plot.title <- "15 chromatin states from 10 brain types" -->

<!-- GetChromHMMPlotAll(df, cbPalette, plot.title) -->
<!-- ``` -->
