***
# __Identification of independent meQTLs: Global approach__
***

## Methodology

The number of independent SNPs was identified for the entire set of meQTLs obtained from the R package "MatrixEQTL" and which passed nominal significance threshold _p < 0.05_ by LD pruning and clumping the SNPs using the “clump” command in PLINK (using distance < __200 kb__ and __r2 > 0.2__ as setting). Each SNP forms a SNP bin, by aggregating all other SNPs into bins by the tag SNP at r2 > 0.2 and distance < 200 kb, such that all SNPs within a given bin were correlated to the current SNP, but to any other SNP. 

```{sh include = F, eval = F}
snp_dir=/binder/mgp/datasets/2020_DexStim_Array_Human/snps
out_dir=/home/ahryhorzhevska/mpip/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/ind_me-qtl_cis

treatment=veh

echo "SNP P" > ind_me-qtl_cis/${treatment}/me-qtl_cis_${treatment}_snps_for_ld_clump.csv

awk 'NR != 1 {print $1 "\t" $5}' me-qtl_cis_result_${treatment}_fdr_005.csv | uniq >> ind_me-qtl_cis/${treatment}/me-qtl_cis_${treatment}_snps_for_ld_clump.csv

# sed -i '1 i\SNP\tP' ind_me-qtl_cis/me-qtl_cis_${treatment}_snps_for_ld_clump.csv

plink --bfile $snp_dir/Dex_genoData_SNPs --clump ind_me-qtl_cis/${treatment}/me-qtl_cis_${treatment}_snps_for_ld_clump.csv --clump-r2 0.2 --clump-kb 200 --out ind_me-qtl_cis/${treatment}/me-qtl_cis_ind_${treatment}

# Extract the index SNP ID:
awk 'NR!=1 {print $3}' ind_me-qtl_cis/${treatment}/me-qtl_cis_ind_${treatment}.clumped >  ind_me-qtl_cis/${treatment}/me-qtl_cis_ind_${treatment}_valid_snps.txt
```

```{r, include = F, eval = T}
library(data.table)
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/"

meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_dex_fdr_005.csv")
meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_indp_result_veh_fdr_005.csv")
meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_result_delta_fdr_005.csv")

col.names <-  c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")

ind.meqtl.dex.df   <- fread(meqtl.dex.fn, col.names = col.names)
ind.meqtl.veh.df   <- fread(meqtl.veh.fn, col.names = col.names)
ind.meqtl.delta.df <- fread(meqtl.delta.fn, col.names = col.names)
```

```{r, include = F, eval = T}
meqtl.all.full.df <- rbind(ind.meqtl.dex.df[, treatment := "dex"],
                           ind.meqtl.veh.df[, treatment := "veh"],
                           ind.meqtl.delta.df[, treatment := "delta"])

meqtl.all.full.df <- meqtl.all.full.df[!is.na(meqtl.all.full.df$fdr),]
meqtl.all.full.df[["meQTL_ID"]] <- paste(meqtl.all.full.df$SNP, meqtl.all.full.df$CpG_ID, sep = "-")
meqtl.all.full.df <- setDT(meqtl.all.full.df)
```

```{r include = F, eval = T}
freq <- fread("/Users/anastasiia_hry/bio/code/mpip/dex-stim-human-array/data/snps/imputed_qc/from_janine/qc/freq.afreq")
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(ind.meqtl.delta.df$fdr  %>% length(), 
                             ind.meqtl.delta.df$CpG_ID %>% unique() %>% length(),
                             ind.meqtl.delta.df$SNP %>% unique() %>% length()),
                           c(ind.meqtl.dex.df$fdr  %>% length(), 
                             ind.meqtl.dex.df$CpG_ID %>% unique() %>% length(),
                             ind.meqtl.dex.df$SNP %>% unique() %>% length()),
                           c(ind.meqtl.veh.df$fdr  %>% length(), 
                             ind.meqtl.veh.df$CpG_ID %>% unique() %>% length(),
                             ind.meqtl.veh.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "dex", "veh")
```

```{r out.width = "95%", include = T, eval = T}

cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")

sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
                  stat = "identity", 
                  vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank()) +
  scale_fill_manual(values = cbPalette)
```

## Upset plots{.tabset .tabset-fade .tabset-pills}

### meQTL level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.meqtls <- list(delta = meqtl.all.full.df[treatment == "delta", meQTL_ID], 
                      dex = meqtl.all.full.df[treatment == "dex", meQTL_ID], 
                      veh = meqtl.all.full.df[treatment == "veh", meQTL_ID])

upset(fromList(meqtls.meqtls), 
      sets = c("delta", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, SNP", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c( "#0072B2", "#009E73", "#E69F00"), 
      order.by = "freq")
```

### SNP level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.snps <- list(delta = ind.meqtl.delta.df$SNP, 
                    dex = ind.meqtl.dex.df$SNP, 
                    veh = ind.meqtl.veh.df$SNP)

upset(fromList(meqtls.snps), 
      sets = c("delta", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, SNP", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00"), 
      order.by = "freq")
```

### CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.cpg <- list(delta = ind.meqtl.delta.df$CpG_ID, 
                   dex = ind.meqtl.dex.df$CpG_ID, 
                   veh = ind.meqtl.veh.df$CpG_ID)

upset(fromList(meqtls.cpg), 
      sets = c("delta", "dex", "veh"),
      nsets = 3, 
      nintersects = 10, 
      mainbar.y.label = "Number of intersections, CpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c("#0072B2", "#009E73", "#E69F00"), 
      order.by = "freq",
      )
```

## Scatter plot

```{r include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/integrative/eqtl/04_me-qtl_analysis/util.R")
cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")
```

```{r out.width = "95%", include = T, eval = T}
fdr.thr <- 0.05
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nResult of MatrixEQTL")
GetScatterPlot2(meqtl.all.full.df, plot.title = plot.title)
```

```{r out.width = "95%", include = T, eval = T}
fdr.thr <- 0.05

plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nResult of MatrixEQTL")

cbPalette <- c( "#0072B2", "#009E73", "#E69F00", "#F0E442", "#D55E00", "#CC79A7", "#56B4E9", "#999999")

ggplot(meqtl.all.full.df, aes(x = `t-stat`, y = -log10(fdr), shape = treatment, color = treatment)) +
  geom_point(alpha = 1.5, size = 1.2) +
  scale_x_continuous(labels = scientific) +
  geom_vline(xintercept = 0, colour = "#990000", linetype = "dashed") + 
  geom_hline(yintercept = -log10(fdr.thr), colour = "#990000", linetype = "dashed") +
  theme(legend.position = "bottom", 
        legend.title = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 8)) +
  labs(x = "t-stat, MatrixEQTL",
       title = plot.title) +
  scale_colour_manual(values = cbPalette)
```

# __Fold Change. Manual Calculation__

## Methodology 

For **veh** and **dex**, calculate:

1. mean beta (carrier of AA)

2. mean beta (carrier of BB)

3. mean beta AA - mean beta BB

- beta = actual methylation values after norm + batch correction (not residuals)

- to get AA (homozyg. major allele) & BB (homozyg. minor allele), use the SNP data, which was used as an input for MatrixEQTL. 

For each meQTL SNP, 

a) extract the individuals with allele 0 (AA) and for those  calculate mean in beta

b) extract the individuals with allele 2 (BB) and for those calculate mean in beta

For **delta**:

Simply calculate the change of beta via beta value of dex - veh


<!-- ## Allele Frequencies -->

<!-- Allele frequencies for each SNP were computed using PLINK and SNPs with minor allele frequencies < 0.2 were excluded -->


## CpGs with low variability

<!-- To eliminate CpGs with low variability across patients, we kept sites with a median absolute deviation (MAD) higher than 0.9. -->


```{r include = F, eval = T}
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/"

methyl.change.dex.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_methyl_change_dex_fdr_005.csv")
methyl.change.dex.df <- fread(methyl.change.dex.fn)

methyl.change.veh.fn <- paste0(out.dir.pre, "me-qtl_cis_indp_methyl_change_veh_fdr_005.csv")
methyl.change.veh.df <- fread(methyl.change.veh.fn)

methyl.change.dex.df <- left_join(methyl.change.dex.df, ind.meqtl.dex.df)
methyl.change.veh.df <- left_join(methyl.change.veh.df, ind.meqtl.veh.df)

methyl.change.df     <- rbind(methyl.change.dex.df , methyl.change.veh.df)

lmem.rslt.fn <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/01_lmem_dnam/dnam_lmem_svs_pcs_rslt.txt'
lmem.rslt    <- fread(lmem.rslt.fn)
ind.meqtl.delta.df.2 <- left_join(ind.meqtl.delta.df, lmem.rslt, by = c("CpG_ID" = "PROBE_ID")) %>%
  select(colnames(ind.meqtl.delta.df), "FC")
colnames(ind.meqtl.delta.df.2)[ncol(ind.meqtl.delta.df.2)] <- "methyl_change"
# ind.meqtl.delta.df.2[["treatment"]] <- "delta"

methyl.change.df.plt <- rbind(methyl.change.df[, c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr", "treatment", "methyl_change")],
                              ind.meqtl.delta.df.2[, c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr", "treatment", "methyl_change")])

methyl.change.df.plt[["meQTL_ID"]] <- paste(methyl.change.df.plt$SNP, methyl.change.df.plt$CpG_ID, sep = "-")
methyl.change.df.plt$methyl_change <- as.numeric(methyl.change.df.plt$methyl_change)

```
### DNAm Metrics Density Plots {.tabset .tabset-fade .tabset-pills}

#### MAD Score

```{r out.width = "95%", include = T, eval = T}
library(ggplot2)

mad.df <- fread("~/bio/code/mpip/dex-stim-human-array/output/data/methylation/mad_score_dnam_beta_values.csv")

ggplot(mad.df, aes(x = MAD, fill = treatment, color = treatment)) +
  geom_density(alpha = 0.2, size = 1) +
  theme(legend.position = "bottom", 
        legend.title = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 8)) +
  labs(title = "DNAm MAD Score Density Plot. Based on beta-values") +
  scale_colour_manual(values = cbPalette[2:3])

```

## Scatter Plots


```{r out.width = "95%", include = T, eval = T}
fdr.thr <- 0.05
plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nManual calculation. No filters")
GetScatterPlot3(methyl.change.df.plt, plot.title = plot.title)
```
```{r out.width = "95%", include = T, eval = T}
library(ggplot2)

fdr.thr <- 0.05
mad.thr <- 75
maf.thr <- 0.1

mad.cpgs.veh <- mad.df[mad.df$perc >= mad.thr & mad.df$treatment == "veh", CpG_ID]
mad.cpgs.dex <- mad.df[mad.df$perc >= mad.thr & mad.df$treatment == "dex", CpG_ID]
mad.cpgs.delta <- intersect(mad.cpgs.veh, mad.cpgs.dex) 

methyl.change.df.plt.filt <- data.frame(rbind(
  methyl.change.df.plt[treatment == "veh" & CpG_ID %in% mad.cpgs.veh],
  methyl.change.df.plt[treatment == "dex" & CpG_ID %in% mad.cpgs.dex],
  methyl.change.df.plt[treatment == "delta" & CpG_ID %in% mad.cpgs.delta]))

methyl.change.df.plt.filt <- left_join(methyl.change.df.plt.filt, freq, by = c("SNP" = "ID")) %>% setDT()
methyl.change.df.plt.filt <- methyl.change.df.plt.filt[ALT_FREQS >= maf.thr ]

plot.title <- paste0("Independent and significant at FDR <= ", fdr.thr, " meQTLs.",
                     "\nManual calculation. MAD >= ", mad.thr, "% and MAF >= ", maf.thr)
GetScatterPlot3(methyl.change.df.plt.filt, plot.title = plot.title)
```
