---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 600,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)
library(ggplot2)
library(scales)
library(corrplot)
library(factoextra)
# library(viridis)
```

# Introduction
***
To explore the cis regulatory effects of SNPs on DNA methylation, the meQTL analysis was performed using a linear regression implemented in the R package "MatrixEQTL".

In the meQTL study, the imputed QC __SNP data__ (no LD pruning, the MHC regions are included) were used.
The data includes __3'957'338 SNPs__.

The __DNAm data__ underwent the folowing QC analysis steps:

```{r dnam_qc_roadmap,  out.width = '100%'}
# knitr::include_graphics("~/bio/code/mpip/dex-stim-human-array/code/integrative/Rmds/img/dnam_qc_roadmap.png")
knitr::include_graphics("../Rmds/img/dnam_qc_roadmap.png")
```

As a result, __740'358 CpGs__ were included in the meQTL analysis.

The correction for treatment, sex, case / control status, age, BMI,  smoking score, first two principal components (PCs) for population stratification and first three significant Surrogate Variables (SVs) was done in a cis-window of Â±1M bp using the linear regression model of Matrix eQTL. 

Delat (differences between baseline and dex-stimulation) was donr based on residuals following the algorithm:

  1. lm_treatment <- lm(beta.mtrx[cpg, ] ~ Sex + Age + BMI + Status + Smoking_Score +
                                           DNAm_SV_treatment{1-3} + PC{1-2})

  2. res_treatment <- residuals(lm_treatment)
  
  3. res_delta <- res_dex - res_veh
 
  4. compute meQTLs using res_delta from (3) adjusting for sex, case / control status, age, BMI,  smoking score, first two principal components (PCs) for population stratification.

The MatrixEQTL meQTL analysis led to 

  + __1'970'917'769__ performed tests

  + __123'399'645__ dex significant (p < 0.05) cis-meQTLs

  + __123'574'934__ veh significant (p < 0.05) cis-meQTLs

  + __100,206,545__ delta significant (p < 0.05) cis-meQTLs

The false discovery rate (FDR) of 5% was chosen. 

```{sh, include = F, eval = F}
meqtl_rslt_dir="~/mpip/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL"
awk 'NR == 1 || $6 < 0.05' me-qtl_cis_result_dex.csv > me-qtl_cis_result_dex_fdr_005.csv
```

```{r, include = F, eval = T}
src.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/data/integrative/matrixEQTL/"
# out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/geno_imputed_qc_ld/"
out.dir.pre  <- "~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/"

cpg.loc.fn  <- paste0(src.dir.pre, "cpg_locations.csv")
snp.loc.fn  <- paste0(src.dir.pre, "snp_locations.csv")
ensg.loc.fn <- paste0(src.dir.pre, "ensg_locations.csv")

cpg.loc  <- fread(cpg.loc.fn)
snp.loc  <- fread(snp.loc.fn)
# ensg.loc <- fread(ensg.loc.fn)

# snp.bim  <- fread("~/bio/code/mpip/dex-stim-human-array/data/snps/final_imputed_qc_snps/mtrx/dex_geno_imputed_ld.bim")
snp.bim  <- fread("~/bio/code/mpip/dex-stim-human-array/data/snps/imputed_qc/from_janine/qc/Dex_genoData_SNPs.bim")
colnames(snp.bim) <- c("chr", "SNP", "pos_morgans", "pos", "snp_allele1", "snp_allele2")

meqtl.dex.fn   <- paste0(out.dir.pre, "me-qtl_cis_result_dex_fdr_005.csv")
meqtl.veh.fn   <- paste0(out.dir.pre, "me-qtl_cis_result_veh_fdr_005.csv")
meqtl.delta.fn <- paste0(out.dir.pre, "me-qtl_cis_result_delta_fdr_005.csv")
# meqtl.delta.fn <- paste0("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/geno_imputed_qc_ld/", "me-qtl_cis_result_delta.csv")

meqtl.dex.df   <- fread(meqtl.dex.fn)
meqtl.veh.df   <- fread(meqtl.veh.fn)
meqtl.delta.df <- fread(meqtl.delta.fn)
```

```{r, include = F, eval = T}
GetFullmeQTLdf <- function(meqtl.df, fdr.thr = 0.05){
  colnames(meqtl.df) <- c("SNP", "CpG_ID", "beta", "t-stat", "p-value", "fdr")
  
  meqtl.fltr.df <- meqtl.df[fdr < fdr.thr, ] %>% unique()
  
  meqtl.full.df <- left_join(meqtl.fltr.df, snp.bim) %>% mutate(pos_snp = pos) %>% select(-c(pos, pos_morgans))
  meqtl.full.df <- left_join(meqtl.full.df, cpg.loc[, .(CpG_ID, chr, pos)]) %>% 
    mutate(pos_cpg = pos) %>% select(-pos)
  meqtl.full.df <- meqtl.full.df[, dist := pos_snp - pos_cpg]  
  meqtl.full.df <- meqtl.full.df[, strand := ifelse(dist < 0, "-", "+")]
  meqtl.full.df <- meqtl.full.df[, dist := abs(dist)]
  
  meqtl.full.df
}
```

```{r, include = F, eval = F}
meqtl.rds <- readRDS("~/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL/me-eqtl_matrx_delta.RDS")
plot(meqtl.rds)
```

```{r, include = F, eval = F}
fdr.thr <- 1

meqtl.delta.no.fdr.df <- GetFullmeQTLdf(meqtl.delta.df, fdr.thr = fdr.thr)
meqtl.dex.no.fdr.df   <- GetFullmeQTLdf(meqtl.dex.df, fdr.thr = fdr.thr)
meqtl.veh.no.fdr.df   <- GetFullmeQTLdf(meqtl.veh.df, fdr.thr = fdr.thr)
```

```{r, include = F, eval = T}
# head(meqtl.df)
fdr.thr  <- 0.05

meqtl.delta.full.df <- GetFullmeQTLdf(meqtl.delta.df, fdr.thr)
meqtl.dex.full.df   <- GetFullmeQTLdf(meqtl.dex.df, fdr.thr)
meqtl.veh.full.df   <- GetFullmeQTLdf(meqtl.veh.df, fdr.thr)
```

```{r, include = F, eval = T}
meqtl.all.full.df <- rbind(meqtl.dex.full.df[, treatment := "dex"],
                           meqtl.veh.full.df[, treatment := "veh"],
                           meqtl.delta.full.df[, treatment := "delta"])
```

<!-- # MANHATTAN PLOT -->
```{r, include = F, eval = F}
library(stringr)
library(QCEWAS)

a <- str_split_fixed(me.add.cis$gene, "-", 2)
me.add.cis$chr <- substring(a[,1],4,length(a[,1]))
me.add.cis$POS <- a[,2]
colnames(me.add.cis) <- c("snps" ,'gene',"stats" ,"P_VAL" , "fdr", 'beta',"CHR", 'POS')
me.add.cis$CHR <- as.numeric(me.add.cis$CHR)
me.add.cis$POS <- as.numeric(me.add.cis$POS)

trace("EWAS_plots",edit=TRUE)
meqtl.delta.no.fdr.df[, P_VAL := `p-value`]
meqtl.delta.no.fdr.df[, CHR := chr]
meqtl.delta.no.fdr.df[, MAPINFO := pos_cpg]
EWAS_plots(meqtl.delta.no.fdr.df,
           plot_QQ = FALSE,
           plot_Man = TRUE,
           plot_cutoff_p = 1,
           plot_QQ_bands = FALSE,
           high_quality_plots = FALSE,
           save_name = paste("~/bio/code/mpip/dex-stim-human-array/output/plots/meqtl_delta_mth_plt")
           )
```

# Hits

Check the number of found eQTLs, unique CpGs and SNPs in all datasets.

<!-- ## Before filtering -->

```{r, include = F, eval = F}
hits.df <- data.table(c(meqtl.delta.no.fdr.df$fdr  %>% length(), 
                        meqtl.delta.no.fdr.df$CpG_ID %>% unique() %>% length(),
                        meqtl.delta.no.fdr.df$SNP %>% unique() %>% length()),
                           c(meqtl.dex.no.fdr.df$fdr  %>% length(), 
                             meqtl.dex.no.fdr.df$CpG_ID %>% unique() %>% length(),
                             meqtl.dex.no.fdr.df$SNP %>% unique() %>% length()),
                           c(meqtl.veh.no.fdr.df$fdr  %>% length(), 
                             meqtl.veh.no.fdr.df$CpG_ID %>% unique() %>% length(),
                             meqtl.veh.no.fdr.df$SNP %>% unique() %>% length()))

hits.df <- as.data.frame(hits.df)
rownames(hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(hits.df) <- c("delta", "dex", "veh")
```

```{r out.width = "95%", include = F, eval = F}
hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
   # geom_violin(width = 0.4) +
    geom_col() +
    geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
              stat = "identity", 
              vjust = -0.5, size = 3) +
    # scale_y_continuous(labels = unit_format(unit = "M", scale = 1e-6)) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank())
```

## Significant Hits, FDR < 0.05

```{r, include = F, eval = T}
sign.hits.df <- data.table(c(meqtl.delta.full.df$fdr  %>% length(), 
                             meqtl.delta.full.df$CpG_ID %>% unique() %>% length(),
                             meqtl.delta.full.df$SNP %>% unique() %>% length()),
                           c(meqtl.dex.full.df$fdr  %>% length(), 
                             meqtl.dex.full.df$CpG_ID %>% unique() %>% length(),
                             meqtl.dex.full.df$SNP %>% unique() %>% length()),
                           c(meqtl.veh.full.df$fdr  %>% length(), 
                             meqtl.veh.full.df$CpG_ID %>% unique() %>% length(),
                             meqtl.veh.full.df$SNP %>% unique() %>% length()))

sign.hits.df <- as.data.frame(sign.hits.df)
rownames(sign.hits.df) <- c("cis-meQTLs", "CpG", "SNP")
colnames(sign.hits.df) <- c("delta", "dex", "veh")
```

```{r out.width = "95%", include = T, eval = T}
sign.hits.df %>% as.matrix() %>% reshape2::melt(var.names = c("ID", "Treatment"), value.name = "cnt") %>%
  ggplot(aes(y = cnt, x = Var2, fill = as.factor(as.numeric(Var2)))) +
   # geom_violin(width = 0.4) +
    geom_col() +
        geom_text(aes(label = comma(cnt, accuracy = 1L), y = cnt), 
              stat = "identity", 
              vjust = -0.2, size = 2.5) + 
    scale_y_continuous(labels = scientific) +
    # scale_fill_viridis(discrete = TRUE, alpha = 0.9, option = "C") +
    labs(title = " ", y = "Count", x = " ") + 
    facet_wrap(~ Var1, ncol = 3) +
    theme(legend.position = "none", 
          legend.title = element_blank(),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8),
          axis.title.x = element_blank())
```

# Upset plots{.tabset .tabset-fade .tabset-pills}

## SNP level
  
```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.snps <- list(delta = meqtl.delta.full.df$SNP, dex = meqtl.dex.full.df$SNP, veh = meqtl.veh.full.df$SNP)

upset(fromList(meqtls.snps), 
      sets = c("delta", "dex", "veh"),
      nsets = 3, 
      nintersects = 20, 
      mainbar.y.label = "Number of intersections, SNP", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c(2, 3, 4), 
      order.by = "freq")
```

## CpG level

```{r out.width = "95%", include = T, eval = T}
library(UpSetR)

meqtls.cpg <- list(delta = meqtl.delta.full.df$CpG_ID, dex = meqtl.dex.full.df$CpG_ID, veh = meqtl.veh.full.df$CpG_ID)

upset(fromList(meqtls.cpg), 
      sets = c("delta", "dex", "veh"),
      nsets = 3, 
      nintersects = 10, 
      mainbar.y.label = "Number of intersections, CpG", 
      text.scale = 1, 
      keep.order = T,
      sets.bar.color = c(2, 3, 4), 
      order.by = "freq",
      )
```

# Scatter plot {.tabset .tabset-fade .tabset-pills}

## all

```{r out.width = "95%", include = T, eval = T}
  ggplot(meqtl.all.full.df, aes(x = beta, y = -log10(fdr), shape = treatment, color = treatment)) +
    geom_point(alpha = 1.5, size = 1.2) +
    scale_x_continuous(labels = scientific) +
    geom_vline(xintercept = 0, colour = "#990000", linetype = "dashed") + 
    geom_hline(yintercept = -log10(fdr.thr), colour = "#990000", linetype = "dashed") +
    # scale_y_continuous(trans = 'log10')+
    # labs(title = " ", y = "", x = " ") + 
    theme(legend.position = "bottom", 
          legend.title = element_blank(),
         #  panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8))
```

## delta

```{r out.width = "95%", include = T, eval = T}

  ggplot(meqtl.delta.full.df, aes(x = beta, y = -log10(fdr))) +
    geom_point(alpha = 1.5, size = 1.2, shape = 16, color = "red") +
    scale_x_continuous(labels = scientific) +
    geom_vline(xintercept = 0, colour = "#990000", linetype = "dashed") + 
    geom_hline(yintercept = -log10(fdr.thr), colour = "#990000", linetype = "dashed") +
    # scale_y_continuous(trans = 'log10')+
    # labs(title = " ", y = "", x = " ") + 
    theme(legend.position = "none", 
          legend.title = element_blank(),
         #  panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8))
```

## dex

```{r out.width = "95%", include = T, eval = T}

  ggplot(meqtl.dex.full.df, aes(x = beta, y = -log10(fdr))) +
    geom_point(alpha = 1.5, size = 1.2, shape = 17, color = "green") +
    scale_x_continuous(labels = scientific) +
    geom_vline(xintercept = 0, colour = "#990000", linetype = "dashed") + 
    geom_hline(yintercept = -log10(fdr.thr), colour = "#990000", linetype = "dashed") +
    # scale_y_continuous(trans = 'log10')+
    # labs(title = " ", y = "", x = " ") + 
    theme(legend.position = "none", 
          legend.title = element_blank(),
         #  panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8))
```

## veh

```{r out.width = "95%", include = T, eval = T}

  ggplot(meqtl.veh.full.df, aes(x = beta, y = -log10(fdr))) +
    geom_point(alpha = 1.5, size = 1.2, shape = 18, color = "blue") +
    scale_x_continuous(labels = scientific) +
    geom_vline(xintercept = 0, colour = "#990000", linetype = "dashed") + 
    geom_hline(yintercept = -log10(fdr.thr), colour = "#990000", linetype = "dashed") +
    # scale_y_continuous(trans = 'log10')+
    # labs(title = " ", y = "", x = " ") + 
    theme(legend.position = "none", 
          legend.title = element_blank(),
         #  panel.grid.major = element_blank(),
          panel.background = element_blank(),
          plot.title = element_text(size = 10),
          axis.title = element_text(size = 8))
```

# Distances

```{r out.width = "95%", include = F, eval = F}
ggplot(meqtl.delta.full.df, aes(x = dist)) + geom_histogram(alpha = .9)
ggplot(meqtl.dex.full.df, aes(x = dist)) + geom_histogram(alpha = .9)
ggplot(meqtl.veh.full.df, aes(x = dist)) + geom_histogram(alpha = .9)
```

```{r out.width = "95%", include = T, eval = T}
ggplot(meqtl.all.full.df, aes(x = dist, fill = treatment)) + 
  geom_density(alpha = 0.2, aes(color = treatment)) +
  theme(legend.position = c(.9,.9), 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 10)) +
  scale_x_continuous(labels = scientific) +
  scale_y_continuous(labels = scientific) +
  labs(title = paste0("Distribution of cis-meQTLs, FDR = ", fdr.thr), 
       x = "Distance of CpG meQTL to SNP", 
       y = "Density") 
```

```{r out.width = "95%", include = F, eval = T}
ggplot(meqtl.all.full.df, aes(x = dist, fill = treatment)) + 
  geom_histogram(alpha = 0.1, aes(color = treatment)) +
  theme(legend.position = c(.9,.9), 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 10)) +
  scale_x_continuous(breaks = c(0, 1e2, 1e3, 1e4, 1e5, 1e6), labels = scientific) +
  scale_y_continuous(labels = scientific) +
  labs(title = paste0("Distribution of cis-meQTLs, FDR = ", fdr.thr), 
       x = "Distance of CpG meQTL to SNP", 
       y = "Density") 
```

```{r out.width = "100%", include = T, eval = T}
library(mltools)
library(ggrepel)

bins.df <- data.frame(DistanceInterval = bin_data(data.table::data.table(meqtl.all.full.df), binCol = "dist", bins = c(0, 1, 200, 1e3, 1e4+1, 1e5+1, 1e6+1, 1e7+1, Inf), returnDT = F))
levels(bins.df$DistanceInterval) <- c("0", "(0; 200]", "(200; 1e3]", "(1e3; 1e4]", "(1e4; 1e5]", "(1e5; 1e6]", "(1e6; 1e7]", "[1e7; Inf)")
bins.df <- cbind(bins.df, treatment = meqtl.all.full.df$treatment)

ggplot(bins.df, aes(DistanceInterval, fill = treatment)) + 
  geom_bar(position = position_dodge()) +
  stat_count(geom = "text", 
             aes(label = comma(..count.., accuracy = 1L)),
             position = position_dodge(1),  vjust = -1, colour = "black", cex = 2) +
  theme(legend.position = c(.1,.9), 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        axis.text.y = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 10)) +
  # scale_y_continuous(breaks = c(0, 1500, 10000, 15000, 20000, 250000, 30000), labels = scientific) +
  labs(title = paste0("Distribution of cis-meQTLs, FDR = ", fdr.thr), 
       x = "Distance Interval, bp", 
       y = "No. cis-meQTLs") 
```

# Distribution of CpGs per SNPs {.tabset .tabset-fade .tabset-pills}

## delta

```{r out.width = "100%", include = T, eval = T}

# plt.df <- data.frame(rbind(cbind(Frequency = plyr::count(meqtl.delta.full.df$SNP)$freq, treatment = "delta"),
#                            cbind(Frequency = plyr::count(meqtl.dex.full.df$SNP)$freq, treatment = "dex"),
#                            cbind(Frequency = plyr::count(meqtl.veh.full.df$SNP)$freq, treatment = "veh")))

plt.df <- data.frame(cbind(Frequency = plyr::count(meqtl.delta.full.df$SNP)$freq, treatment = "delta"))
plt.df$Frequency <- as.numeric(plt.df$Frequency)

nr.bins <- max(plt.df$Frequency)

ggplot(plt.df, aes(x = Frequency)) + 
  geom_histogram(color = "red", fill = "red", bins =  nr.bins, alpha = 0.5) +
  stat_count(geom = "text", 
             aes(label = comma(..count.., accuracy = 1L)),
             position = position_dodge(1), angle = 0,  vjust = -0.5, hjust = 0.5, colour = "black", cex = 3) +
  theme(legend.position = "none", 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 10),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  scale_x_continuous(breaks = seq(1, nr.bins, 1)) +
  scale_y_continuous(labels = scientific) +
  labs(title = paste0("Distribution of CpGs per SNPs (cis), FDR < ", fdr.thr), 
       y = "Density", 
       x = "Nr of CpGs") +
  facet_wrap(~ treatment, ncol = 1)
```

```{r out.width = "95%", include = F, eval = F}
hist(plyr::count(meqtl.delta.full.df$SNP)$freq, 
     breaks = max(plyr::count(meqtl.delta.full.df$SNP)$freq) + 1,
     main = paste0("Distribution of CpGs per SNPs (delta) fdr < ", fdr.thr),
     xlab = 'Nr of CpGs')
```

## dex

```{r out.width = "100%", include = T, eval = T}
plt.df <- data.frame(cbind(Frequency = plyr::count(meqtl.dex.full.df$SNP)$freq, treatment = "dex"))
plt.df$Frequency <- as.numeric(plt.df$Frequency)

nr.bins <- max(plt.df$Frequency)

ggplot(plt.df, aes(x = Frequency)) + 
  geom_histogram(color = "green", fill = "green", bins =  nr.bins, alpha = 0.5) +
  stat_count(geom = "text", 
             aes(label = comma(..count.., accuracy = 1L)),
             position = position_dodge(1), angle = 80,  hjust = -0.2, colour = "black", cex = 2) +
  theme(legend.position = "none", 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 10),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 6)) +
  scale_x_continuous(breaks = seq(1, nr.bins, 1)) +
  scale_y_continuous(labels = scientific) +
  labs(title = paste0("Distribution of CpGs per SNPs (cis), FDR < ", fdr.thr), 
       y = "Density", 
       x = "Nr of CpGs") +
  facet_wrap(~ treatment, ncol = 1)
```

## veh

```{r out.width = "100%", include = T, eval = T}
plt.df <- data.frame(cbind(Frequency = plyr::count(meqtl.veh.full.df$SNP)$freq, treatment = "veh"))
plt.df$Frequency <- as.numeric(plt.df$Frequency)

nr.bins <- max(plt.df$Frequency)

ggplot(plt.df, aes(x = Frequency)) + 
  geom_histogram(color = "blue", fill = "blue", bins =  nr.bins, alpha = 0.5) +
  stat_count(geom = "text", 
             aes(label = comma(..count.., accuracy = 1L)),
             position = position_dodge(1), angle = 80,  hjust = -0.2, colour = "black", cex = 2) +
  theme(legend.position = "none", 
        legend.title = element_blank(),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 10),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 6)) +
  scale_x_continuous(breaks = seq(1, nr.bins, 1)) +
  scale_y_continuous(labels = scientific) +
  labs(title = paste0("Distribution of CpGs per SNPs (cis), FDR < ", fdr.thr), 
       y = "Density", 
       x = "Nr of CpGs") +
  facet_wrap(~ treatment, ncol = 1)
```