---
title: "cis-meQTLs Analysis"
subtitle: "Dexamethasone-Stimulated Human Array Project"
author: 
  - name: Anastasiia Hryhorzhevska
    email: anastasiia_hry@psych.mpg.de
    url: https://github.com/ahryho
    affiliation: Max Planck Institute of Psychiatry
    affiliation_url: https://www.psych.mpg.de/2664393/medizinische-genomforschung
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 600,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
library(DT)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)
library(ggplot2)
library(scales)
library(corrplot)
library(factoextra)
library(viridis)
```

# Introduction
***
To explore the cis regulatory effects of SNPs on DNA methylation, the meQTL analysis was performed using a linear regression implemented in the R package "MatrixEQTL".

In the meQTL study, the imputed QC __SNP data__ (no LD pruning, the MHC regions are included) were used.
The data includes __3'957'338 SNPs__.

The __DNAm data__ underwent the folowing QC analysis steps:

```{r dnam_qc_roadmap,  out.width = '100%'}
# knitr::include_graphics("~/bio/code/mpip/dex-stim-human-array/code/integrative/Rmds/img/dnam_qc_roadmap.png")
knitr::include_graphics("../../Rmds/img/dnam_qc_roadmap.png")
```

As a result, __740'358 CpGs__ were included in the meQTL analysis.

The correction for treatment, sex, case / control status, age, BMI,  smoking score, first two principal components (PCs) for population stratification and first three significant Surrogate Variables (SVs) was done in a cis-window of Â±1M bp using the linear regression model of Matrix eQTL. 

Delta (differences between baseline and dex-stimulation) was done based on residuals following the algorithm:

  1. lm_treatment <- lm(beta.mtrx[cpg, ] ~ Sex + Age + BMI + Status + Smoking_Score +
                                           DNAm_SV_treatment{1-3} + PC{1-2})

  2. res_treatment <- residuals(lm_treatment)
  
  3. res_delta <- res_dex - res_veh
 
  4. compute meQTLs using res_delta from (3) adjusting for sex, case / control status, age, BMI,  smoking score, first two principal components (PCs) for population stratification.

The MatrixEQTL meQTL analysis led to 

  + __1'970'917'769__ performed tests

  + __123'399'645__ dex significant (p < 0.05) cis-meQTLs

  + __123'574'934__ veh significant (p < 0.05) cis-meQTLs

  + __100,206,545__ delta significant (p < 0.05) cis-meQTLs

The false discovery rate (FDR) of 5% was chosen. 

```{sh, include = F, eval = F}
meqtl_rslt_dir="~/mpip/bio/code/mpip/dex-stim-human-array/output/data/integrative/matrixEQTL"
awk 'NR == 1 || $6 < 0.05' me-qtl_cis_result_dex.csv > me-qtl_cis_result_dex_fdr_005.csv
```

```{r child = c('04_01_me-qtl_analysis.Rmd', '04_03_independent_meQTLs.Rmd')}
# child = c('04_03_independent_meQTLs.Rmd')
# child = c('04_01_me-qtl_analysis.Rmd', '04_02_independent_meQTLs.Rmd')
```