---
title: "Integrative Analysis"
author: "Anastasiia Hryhorzhevska"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    css: style.css
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
options(digits = 4, width = 100, stringsAsFactors = T)
knitr::opts_chunk$set(echo = TRUE,
                      tidy.opts = list(width.cutoff = 100),
                      tidy=TRUE,
                      fig.pos = "H",
                      dpi = 600,
                      warning = FALSE, 
                      message = FALSE, 
                      cache = TRUE, 
                      cache.lazy = FALSE)

# define DT as the standard printing method for data.frame
library(knitr)
knit_print.data.frame <- function(x, ...) {
  knit_print(DT::datatable(x,
                           filter="top",
                           rownames = FALSE,
                           extensions = "FixedColumns",
                           options = list(
                             scrollX = TRUE,
                             pageLength = 5
                             
                           )), ...)
}

registerS3method("knit_print", "data.frame", knit_print.data.frame)

library(data.table)
library(dplyr)
library(ggplot2)
library(corrplot)
library(factoextra)
```

# Introduction

```{r, include = F, eval = T}
# source("~/bio/code/mpip/dex-stim-human-array/code/methylation/02_dma/01_dmp/funcitons.R")
```

1. The multiomics data:
    + Methylation
    + Gene-expression
    + Genotype
    + Bio layer, or phenotype here

For each omic layer, there is different number of observation. The first four section provides an overview and basic statistics for each layer. The last section - basic integrative analysis

# Phenotype

```{r, include = F, eval = T}
pheno.full.fn <- paste0("~/bio/code/mpip/dex-stim-human-array/data/pheno/pheno_full_for_kimono.csv") 
pheno.full    <- read.csv2(pheno.full.fn) %>% setDT()
```

Dex-Stim Porject includes
```{r, include = F, eval = F}
dim(pheno.full) 
colnames(pheno.full)
```

# Methylation 

The roadmap for qc methylation:

```{r dnam_qc_roadmap,  out.width = '100%'}
knitr::include_graphics("img/dnam_qc_roadmap.png")
```

<!-- <center> -->
<!-- ![](~/bio/code/mpip/dex-stim-human-array/data/methylation/img/dnam_qc_roadmap.png){width='80%'} -->

1. Methylation data:
  + BCC: CD8T, CD4T, NK, Bcell, Mono, Gran - cellular proportion estimates from Houseman algorithm
  + 62 sugnificant Surrogate Variables (SV)
  + DNAm Age- age estimated from DNAm data using Hovarth algorithm
  + Smoking Score - smoking score estimated from DNAm data using Elliot's CpGs

## SVA

### Correlation {.tabset .tabset-fade .tabset-pills}

```{r, include = F, eval = T}
pheno      <- pheno.full[Include == 1]
```

<!-- Calculate first 10 DNAm PCs -->

```{r compute-dnam-pcs, include = F, eval = F}
library(lumi)

beta.mtrx.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_beta_combat_mtrx.rds"
beta.mtrx    <- readRDS(beta.mtrx.fn)

beta.mtrx   <- beta.mtrx[, match(pheno$DNAm_ID, colnames(beta.mtrx))]
mval.mtrx   <- beta2m(beta.mtrx)
t.mval.mtrx <- t(mval.mtrx)

pal <- brewer.pal(8, "Set2")

pca.obj <- prcomp(mval.mtrx, retx = F, center = T, scale. = F, rank. = 10)  
# pca.obj.retx <- prcomp(mval.mtrx, retx = T, center = T, scale. = F, rank. = 10)  

# temp <- pca.obj
saveRDS(pca.obj,
        "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_pca.rds")

# write.csv2(dnam.pcs, 
#            "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_top10pcs.csv", 
#            row.names = F, quote = F)
```

```{r load-dnam-pcs, include = F, eval = T}
# dnam.pcs.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_top10pcs.csv"
# dnam.pcs    <- read.csv2(dnam.pcs.fn)

pca.obj.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_pca.rds"
pca.obj    <- readRDS(pca.obj.fn)

nr.pcs   <- ncol(pca.obj$rotation)

propvar  <- round(summary(pca.obj)$importance["Proportion of Variance", 1:nr.pcs] * 100, 2 )
cummvar  <- round(summary(pca.obj)$importance["Cumulative Proportion", 1:nr.pcs] * 100, 2) 

dnam.pcs <- data.frame(pca.obj$rotation)
dnam.pcs <- dnam.pcs[match(pheno$DNAm_ID, rownames(dnam.pcs)), ]

dnam.pcs    <- dnam.pcs[match(pheno$DNAm_ID, rownames(dnam.pcs)), paste0("PC", 1:10)] 
colnames(dnam.pcs) <- paste0("DNAm_PC", 1:10)
```

```{r combine-all-data, include = F, eval = T}
bcc.gex  <- pheno[, c("Cort_D1", "ACTH_D1", "Leukos_D1", "Gran_D1", "Mono_D1", "Lymph_D1")]
bcc.gex <- bcc.gex %>% 
  mutate_all(~ifelse(is.na(.), median(., na.rm = TRUE), .))
bcc.meth <- pheno[, c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran")]
sv.gex   <- pheno[, paste0("V", 1:10)] # c("V1", "V2", "V3", "V4", "V5")]
sv.meth  <- pheno[, paste0("DNAm_SV", 1:10)]# c("DNAm_SV1", "DNAm_SV2", "DNAm_SV3")]
snp.pcs  <- pheno[, paste0("PC", 1:5)]

tmp.df <- data.frame(sv.meth, meth = bcc.meth, gex = bcc.gex, gex = sv.gex, snp.pcs, dnam.pcs,
                     Dex = as.numeric(pheno$Dex), Status = as.numeric(pheno$Status), Sex = as.numeric(pheno$Sex), BMI = pheno$BMI_D1, 
                     Age = pheno$Age, DNAm_Age = pheno$mAge_Hovath, DNAm_Smoke = pheno$DNAm_SmokingScore) %>% 
  na.omit() %>% 
  scale()

cor.mtrx <- cor(tmp.df)
```

#### DNAm SVs vs BCCs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 SVs derived from DNAm data with available and derived covariates", include = T, eval = T}
# cor.mtrx.plt <- cor.mtrx[paste0("DNAm_SV", 1:10), -which(colnames(cor.mtrx) %in% c(paste0("DNAm_SV", 1:10), paste0("DNAm_PC", 1:10)))]
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_SV", 1:10), which(colnames(cor.mtrx) %in% c(paste("meth", colnames(bcc.meth), sep = "." ), 
                                                                                  paste("gex", colnames(bcc.gex), sep = "." )))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

#### DNAm SVs vs GEX SVs and SNP PCs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 SVs derived from DNAm data with available and derived covariates", include = T, eval = T}
# cor.mtrx.plt <- cor.mtrx[-which(rownames(cor.mtrx) %in% paste0("DNAm_SV", 1:10)), paste0("DNAm_SV", 1:10)]
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_SV", 1:10), which(colnames(cor.mtrx) %in% c(paste("gex", colnames(sv.gex), sep = "." ),
                                                                                  paste0("PC", 1:5 )))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

#### DNAm SVs vs Bio

```{r out.width = "98%", fig.cap = "Relationship of the top 10 SVs derived from DNAm data with available and derived covariates", include = T, eval = T}
# cor.mtrx.plt <- cor.mtrx[-which(rownames(cor.mtrx) %in% paste0("DNAm_SV", 1:10)), paste0("DNAm_SV", 1:10)]
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_SV", 1:10), c("Dex", "Status", "Sex", "BMI", "Age", "DNAm_Age", "DNAm_Smoke")]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

## PCA

### Correlation {.tabset .tabset-pills}

#### DNAm PCs vs DNAm SVs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 PCs derived from DNAm data with top 10 SVs", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), paste0("DNAm_SV", 1:10)]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

#### DNAm PCs vs BCCs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 PCs derived from DNAm data with available and derived covariates", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), c(paste("meth", colnames(bcc.meth), sep = "." ), 
                                                    paste("gex", colnames(bcc.gex), sep = "." ))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

#### DNAm PCs vs GEX SVs and SNP PCs

```{r out.width = "98%", fig.cap = "Relationship of the top 10 SVs derived from DNAm data with available and derived covariates", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), c(paste("gex", colnames(sv.gex), sep = "." ),
                                                    paste0("PC", 1:5 ))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

#### DNAm PCs vs Bio

```{r out.width = "98%", fig.cap = "Relationship of the top 10 SVs derived from DNAm data with available and derived covariates", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), c("Dex", "Status", "Sex", "BMI", "Age", "DNAm_Age", "DNAm_Smoke")]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

### {-}

### Analysis{.tabset .tabset-pills}

#### Scree plot

```{r out.width = "90%", fig.cap = "Percentage of variance explained by first 10 eigenvalues"}
propvar.df <- data.frame(PC = 1:10, propvar)
ggplot(propvar.df, aes(PC, propvar)) +
  geom_bar(stat = "identity", width = 0.7, fill = "#00688B") +
  geom_line(colour = "red", linetype = "dashed") +
  geom_point(colour = "red") +
  ylab("Percentage variance explained") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

#### Dex

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the DEX treatment separation", include = T, eval = T}
plt.df <- data.frame(scale(dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df, aes(DNAm_PC1, DNAm_PC2, col = Group)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(propvar[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(propvar[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

#### Sex

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the SEX treatment separation", include = T, eval = T}
plt.df <- data.frame(scale(dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df[plt.df$Group == 1,], aes(DNAm_PC6, DNAm_PC7, col = Sex)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC6 (", signif(propvar[6], 2), "%)")) +
  ylab(paste0("PC7 (", signif(propvar[7], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

# Gene-expression


# Genotype


# Integrative Basic Analysis

