***
# DNAm LMEM with DNAm SVs and Genotype PCs
***

## __Roadmap__

Roadmap :

1. Subset data:
    + MAF >= 5%
    + filter only SNPs
    + exclude MHC region
2. LD pruning based on pairwise correlation with the following options:
    + a window of 50 SNPs: to calculate LD between each pair of SNPs
    + remove one of a pair if the LD is greater than 0.2
    + short the window 5 SNPs forward
3. MDS / PCA in PLINK
4. Analysis in R

## __Files location__


<!-- ### Analysis in R -->


```{sh, include = F, eval = F}
scp -C -r -o ProxyCommand="ssh cluser@biocomp.psych.mpg.de nc slurmgate 22" ahryhorzhevska@slurmgate:/binder/mgp/datasets/2020_DexStim_Array_Human/methylation/20_DMA/01_lme_models/dnam_lmem_svs_pcs_rslt.txt output/data/methylation/01_lmem_dnam
```

```{r, include = F, eval = T}
lmem.rslt.fn <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/01_lmem_dnam/dnam_lmem_svs_pcs_rslt.txt'
dmps.anno.fn <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_annotated.csv'

lmem.rslt    <- fread(lmem.rslt.fn)
dmps.anno.df <- fread(dmps.anno.fn)
```

## Tune FDR and FC thresholds

```{r}
pval.list <- c(0.0001, 0.001, 0.01, 0.02, 0.03, 0.04, 0.05)
fc.list   <- c(0, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2)

# beta.list <- c(0, 0.2, 0.3, 0.4, 0.5, 0.7, 0.1)

est.df           <- data.frame(matrix(NaN, nrow = length(pval.list), ncol = length(fc.list)))
colnames(est.df) <- beta.list
rownames(est.df) <- pval.list

for (p.thr in pval.list)
  for (fc.thr in fc.list){
    dmp.sign.df  <- lmem.rslt[abs(FC) > fc.thr & FDR < p.thr, ]
    est.df[as.character(p.thr), as.character(fc.thr)] <- nrow(dmp.sign.df)
  }

est.df <- rownames_to_column(est.df, var = "p-value")
est.df <- est.df %>% pivot_longer(cols = colnames(est.df)[-1], names_to = "FC", values_to = "Nr_DMPs")

est.df$`p-value` <- as.numeric(est.df$`p-value`)
est.df$FC      <- est.df$FC
```

```{r}
ggplot(est.df, aes(`p-value`, Nr_DMPs, color = FC, group = FC)) + 
  geom_point() + geom_line() +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 12),
         axis.text.x = element_text(angle = 0, hjust = 1), 
         legend.position = "bottom") +
  labs(title = " ", y = "Number of DMPs", x = "p-value")
```

```{r }
delta.beta <- 0.01
pval.list <- c(0.001, 0.01, 0.02, 0.03, 0.04, 0.05)
nr.dmps.list <- c()
for (p.thr in pval.list){
  dmp.sign.df  <- lmem.bcc.rslt[abs(FC) > delta.beta & FDR < p.thr, ]
  nr.dmps.list <- append(nr.dmps.list, dim(dmp.sign.df)[1])
}

pval.est <- as.data.frame(cbind(PVAL = pval.list, Nr_DMPs = nr.dmps.list, Pr_DMPs = round(nr.dmps.list / nrow(lmem.rslt) * 100, 2)))

ggplot(pval.est, aes(x = PVAL, y = Nr_DMPs)) + 
  geom_line(linetype="dotted", size=2) +
  geom_point( size = 3) +
  geom_label(aes(label = Nr_DMPs), vjust = 0, nudge_y = 0) +
  theme(legend.position = "bottom")
```

# Get significant dmps

```{r}
beta.list <- myround(c(0.00, 0.01, 0.02, 0.05, 0.10, 0.15, 0.2), digits = 2)
p.thr     <- 9e-08

res <- c()

for (delta.beta in beta.list){
  dmps.sign.df   <- lmem.rslt[abs(FC) > delta.beta & FDR < p.thr,]
  nr.hypermethyl <- nrow(dmps.sign.df[FC > delta.beta & FDR < p.thr,])
  nr.hypomethyl  <- nrow(dmps.sign.df[FC < delta.beta & FDR < p.thr,])
  
  per.hypermethyl <-  myround(nr.hypermethyl / nrow(dmps.sign.df), 4)
  per.hypomethyl <-  myround(nr.hypomethyl / nrow(dmps.sign.df), 4)
  
  res <- rbind(res, c(delta.beta, nrow(dmps.sign.df), nr.hypermethyl, per.hypermethyl, nr.hypomethyl, per.hypomethyl))
}

# datatable(res)
# pander(res)
# kable(res)

res <- data.frame(res, stringsAsFactors = F)
colnames(res) <- c("FC", "NrDMPs", "HyperDMPs", "HyperDMPs (%)", "HypoDMPs", "HypoDMPs (%)") 

res$`HyperDMPs (%)` <- percent(res$`HyperDMPs (%)`)
res$`HypoDMPs (%)` <- percent(res$`HypoDMPs (%)`)
res$NrDMPs <- accounting(res$NrDMPs, big.mark = " ", digits = 0)
res$HyperDMPs <- accounting(res$HyperDMPs, big.mark = " ", digits = 0)
res$HypoDMPs <- accounting(res$HypoDMPs, big.mark = " ", digits = 0)

formattable(res, list(
  area(col = c(`HyperDMPs (%)`)), #~ color_tile("transparent", "lightgray"),
  area(col = c(`HypoDMPs (%)`)) #~ color_tile("transparent", "lightgray") # normalize_bar("lightgray", 0.05) 
))

# formattable(df, list(
#   age = color_tile("white", "orange"),
#   grade = formatter("span", style = x ~ ifelse(x == "A",
#     style(color = "green", font.weight = "bold"), NA)),
#   area(col = c(test1_score, test2_score)) ~ normalize_bar("pink", 0.2),
#   final_score = formatter("span",
#     style = x ~ style(color = ifelse(rank(-x) <= 3, "green", "gray")),
#     x ~ sprintf("%.2f (rank: %02d)", x, rank(-x))),
#   registered = formatter("span",
#     style = x ~ style(color = ifelse(x, "green", "red")),
#     x ~ icontext(ifelse(x, "ok", "remove"), ifelse(x, "Yes", "No")))
# ))
```

```{r}
delta.beta <- 0.0
p.thr      <- 0.05
dmps.sign.df <- lmem.bcc.rslt[abs(FC) > delta.beta & FDR < p.thr,]

nr.hypermethyl <- nrow(dmps.sign.df[FC > delta.beta & FDR < p.thr,])
nr.hypomethyl  <- nrow(dmps.sign.df[FC < delta.beta & FDR < p.thr,])

lmem.bcc.rslt[, "threshold"] <- as.factor(abs(lmem.bcc.rslt$FC) > delta.beta & lmem.bcc.rslt$FDR < p.thr)
cols <- c("TRUE" = "green", "FALSE" = "grey")
# pdf(file = paste0(report.dir, 
#                   "Volcano_plot_model_", mdl, 
#                   "beta_", delta.beta * 100, 
#                   "_p_", p.thr * 100,
#                   ".pdf" ))

ggplot(lmem.bcc.rslt, aes(y = -log10(FDR), x = FC, color = threshold)) +
  geom_point(alpha = .5, size = 1.2) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = delta.beta, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = -delta.beta, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = -log10(p.thr), colour="#990000", linetype="dashed") +
  theme(legend.position = "none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none") +
  ggtitle(paste0("Volcano plot\nP-value threshold = ", p.thr, "\n", nr.hypermethyl, " hypermethyl CpGs\n", nr.hypomethyl, " hypomethyl CpGs"))
```

### __Eigenvalues analysis__ {.tabset .tabset-fade .tabset-pills}

#### Table

```{r out.width="80%"}
pve <- data.frame(PC = 1:nrow(eigenval), eigenval, pve = eigenval/sum(eigenval) * 100)
pve[["CumSum"]] <- cumsum(pve$V1.1)
colnames(pve)   <- c("PC", "Eigenvalue", "pve", "CumSum (%)")
pve
```

#### Plot
```{r out.width="80%", fig.cap = "Percentage of variance explained by eigenvalues"}
ggplot(pve[1:30,], aes(PC, `pve`)) +
  geom_bar(stat = "identity", width = 0.7, fill = "#00688B") +
  ylab("Percentage variance explained") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```


```{r, include = F, eval = T}
# Load pheno data
pheno.methyl.fn <- "~/bio/datasets/pheno/pheno_full_for_kimono.csv"
pheno           <- fread(pheno.methyl.fn)
pheno$Status    <- as.factor(as.character(pheno$Status))
pheno$Sex       <- as.factor(pheno$Sex)

levels(pheno$Status) <- c("Cases", "Controls")
levels(pheno$Sex)    <- c("Male", "Female")
# pheno           <- pheno[!is.na(DNAm_ID),]
# eigenval        <- eigenval[eigenvec$FID %in% pheno$DNA_ID, ]
eigenvec        <- eigenvec[eigenvec$FID %in% pheno$DNA_ID, ]
```

### __Eigenvectors analysis__ {.tabset .tabset-fade .tabset-pills}

#### PLINK

```{r out.width="80%", fig.cap = "Plink MDS Inviduals Map"}
ggplot(data = eigenvec, aes(C1, C2, col = as.factor(SOL))) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  labs(col = "Land") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

#### Clustering {.tabset .tabset-fade .tabset-pills}

##### K-means

```{r out.width="80%", include = F, eval = T}
kmeans.mdl <- kmeans(eigenvec[, -c(1:3)], centers = 2, nstart = 100)
clusters   <- kmeans.mdl$cluster
pca        <- cbind(eigenvec,  clusters)
```


```{r out.width="80%", include = T, eval = T}
ggplot(data = pca, aes(C1, C2, col = as.factor(clusters))) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")

```

##### Hierarhical {.tabset .tabset-fade .tabset-pills}

###### Dendogram

```{r out.width="80%", fig.cap = "Hierarhcical Clustering. Cluter Dendogram", include = T, eval = T}
d      <- dist(eigenvec[, -c(1:3)], method = "euclidean")
hc.mdl <- hclust(d, method = "complete" )
plot(hc.mdl, ylab = "Height", xlab = " ", main = " ")
```

###### Map

```{r include = F, eval = T}
sub.grp  <- as.factor(cutree(hc.mdl, k = 3)) # length(kmeans.mdl$size))
pca      <- left_join(eigenvec, pheno[Dex == 1, .(DNA_ID, Status, SNP_Chip, Sex)], by = c("FID" = "DNA_ID"))
```

```{r out.width="80%", fig.cap = "Hierarchical clustering results by obtained clusters", include = T, eval = T}
ggplot(data = pca, aes(C1, C2, col = sub.grp)) + # as.factor(Status))) + # sub_grp)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

```{r include = F, eval = T, include = F, eval = F}
sub.grp  <- as.factor(cutree(hc.mdl, k = 2)) # length(kmeans.mdl$size))
pca      <- left_join(eigenvec, pheno[Dex == 1, .(DNA_ID, Status, SNP_Chip, Sex)], by = c("FID" = "DNA_ID"))
```

```{r fig.cap = "Hierarchical clustering results by obtained clusters", include = F, eval = F}
ggplot(data = pca, aes(C1, C2, col = sub.grp)) + # as.factor(Status))) + # sub_grp)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

<!-- ### Additional plots {.tabset .tabset-fade .tabset-pills} -->

```{r include = F, eval = T, include = F, eval = T}
pca      <- left_join(eigenvec, pheno[Dex == 1, .(DNA_ID, Status, SNP_Chip, Sex)], by = c("FID" = "DNA_ID"))
```

#### by Chip

```{r out.width="80%", fig.cap = "MDS results by chip"}
ggplot(data = pca, aes(C1, C2, col = SNP_Chip)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

#### by MDD Status 

```{r out.width="80%", fig.cap = "MDS results by status"}
ggplot(data = pca, aes(C1, C2, col = Status)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom") +
  labs(fill = "MDD Status")
```

#### by Sex

```{r out.width="80%"}
ggplot(data = pca, aes(C1, C2, col = Sex)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom") 
```

#### Ethnicity information


```{r out.width="80%", include = T, eval = T}
library(haven)
etnicity.fn <- "/Users/anastasiia_hry/bio/datasets/pheno/MAD_GSK_Phenotype_20191009_for_ah_02072021.sav"
etnicity    <- read_sav(etnicity.fn)

etnicity    <- etnicity[etnicity$NID %in% pheno$DNA_ID, ]
pheno       <- left_join(pheno, etnicity[, c("NID", "S_Land")], by = c("DNA_ID" = "NID"))
eigenvec    <- left_join(eigenvec, etnicity[, c("NID", "S_Land")], by = c("FID" = "NID"))

na.id <- pheno[!(DNA_ID %in% etnicity$NID), "DNA_ID"]
na.id <- na.id[!duplicated(na.id$DNA_ID), ]
```

```{r include = F, eval = F}
write.csv2(na.id,
           "~/bio/datasets/pheno/missed_etnicity_ids.csv",
           row.names = F,
           quote = F)
```

```{r out.width="80%", fig.cap = "MDS result with enthnicity infromation"}
ggplot(data = eigenvec, aes(C1, C2, color = S_Land)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 1, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  labs(col = "Land") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "right")
```

### __Outliers detedtion__ {.tabset .tabset-fade .tabset-pills}

#### 4 SD 

__List of samples which are located 4 mean distances far from the mean of PC1 and PC2__

```{r out.width="80%", include = T, eval = T}
eps     <- 0.01
pc.mean <- mean(eigenvec$C1)
pc.std  <- sd(eigenvec$C1)
cut.off <- 4 * pc.std
lower   <- pc.mean - cut.off + eps
upper   <- pc.mean + cut.off - eps

outliers <- eigenvec[C1 < lower | C1 > upper, FID ]
outliers
```

```{r out.width="80%", fig.cap = "MDS results with labeled outliers"}
ggplot(data = pca, aes(C1, C2, col = Status)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_label_repel(data = pca[pca$FID %in% outliers, ],
                   aes(label = outliers, size = NULL)) + 
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

#### 5 SD

__List of samples which are located 5 mean distances far from the mean of PC1 and PC2__

```{r out.width="80%", include = T, eval = T}
eps     <- 0.01
pc.mean <- mean(eigenvec$C1)
pc.std  <- sd(eigenvec$C1)
cut.off <- 5 * pc.std
lower   <- pc.mean - cut.off + eps
upper   <- pc.mean + cut.off - eps

outliers <- eigenvec[C1 < lower | C1 > upper, FID ]
outliers
```

```{r out.width="80%", fig.cap = "MDS results with labeled outliers"}
ggplot(data = pca, aes(C1, C2, col = Status)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_label_repel(data = pca[pca$FID %in% outliers, ],
                   aes(label = outliers, size = NULL)) + 
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

#### 6 SD

__List of samples which are located 6 mean distances far from the mean of PC1 and PC2__

```{r out.width="80%", include = T, eval = T}
eps     <- 0.01
pc.mean <- mean(eigenvec$C1)
pc.std  <- sd(eigenvec$C1)
cut.off <- 6 * pc.std
lower   <- pc.mean - cut.off + eps
upper   <- pc.mean + cut.off - eps

outliers <- eigenvec[C1 < lower | C1 > upper, FID ]
outliers
```

```{r out.width="80%", fig.cap = "MDS results with labeled outliers"}
ggplot(data = pca, aes(C1, C2, col = Status)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_label_repel(data = pca[pca$FID %in% outliers, ],
                   aes(label = outliers, size = NULL)) + 
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("PC1 (", signif(pve$pve[1], 2), "%)")) +
  ylab(paste0("PC2 (", signif(pve$pve[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "bottom")
```

<!-- ```{r include = F, eval = F} -->
<!-- # Export individuals for dex stim inegrative analysis -->
<!--  -->
<!--  sample.ids <- (pheno[!(DNA_ID %in% outliers),][, .(Sample_ID, DNA_ID, RNA_ID, DNAm_ID)] ) -->
<!--   -->
<!--  write.csv2(sample.ids, -->
<!--              file = "~/bio/code/mpip/dex-stim-human-array/data/sample_ids_for_analysis.csv", -->
<!--              row.names = F, quote = F) -->
<!--  ``` -->
