***
# DMPs
***

## Roadmap

## Files location

## Signififcant DMPs

```{sh, include = F, eval = F}
scp -C -r -o ProxyCommand="ssh cluser@biocomp.psych.mpg.de nc slurmgate 22" ahryhorzhevska@slurmgate:/binder/mgp/datasets/2020_DexStim_Array_Human/methylation/20_DMA/01_lme_models/dnam_lmem_svs_pcs_rslt.txt output/data/methylation/01_lmem_dnam
```

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/methylation/02_dma/01_dmp/funcitons.R")
```

```{r load-lmem-anno, include = F, eval = T}
lmem.rslt.fn <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/01_lmem_dnam/dnam_lmem_svs_pcs_rslt.txt'
anno.fn      <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_annotated.csv'

lmem.rslt    <- fread(lmem.rslt.fn)
anno.df      <- fread(anno.fn)
```


```{r load-pheno, include = F, eval = T}
pheno.full.fn <- paste0("~/bio/code/mpip/dex-stim-human-array/data/pheno/pheno_full_for_kimono.csv") 
pheno.full    <- read.csv2(pheno.full.fn) %>% setDT()

pheno <- pheno.full[Include == 1]

pheno$Status  <- as.factor(pheno$Status)
pheno$Sex     <- as.factor(pheno$Sex)

levels(pheno$Sex) <- c("Male", "Female")
levels(pheno$Status) <- c("Control", "MDD")
```

```{r load-dmps-450k, include = F, eval = T}
dmps.450k.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_q_value01FC002_methylation_450K_CellCode.txt"
dmps.450k    <- read.table(dmps.450k.fn, header = T)
```

```{r dmps-450k-inside, eval = T, include = F}
dmps.450k
```

```{r set-up-params, out.width = "85%", fig.cap = "Model with DNAm BCCs and Genotype PCs", include = T, eval = T}
fdr <- 0.01 #0.05
fc  <- 0.02 #0.01

VolcanoPlot(lmem.rslt, fdr, fc, "DNAm SVs and SNP PCs")
```

<!-- # Get significant dmps -->

<!-- ```{r save-dmps, include = F, eval = F} -->
<!-- dmps.sign.df <- lmem.rslt[abs(FC) > fc & FDR < fdr,] -->

<!-- dmps.sign.anno.df <- anno.df[Name %in% dmps.sign.df$PROBE_ID ] -->
<!-- dmps.sign.anno.df <- left_join(dmps.sign.df, dmps.sign.anno.df, by = c("PROBE_ID" = "Name")) -->

<!-- fwrite(dmps.sign.anno.df,  -->
<!--        "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_fdr01_fc02_anno_full.csv",  -->
<!--        quote = F, row.names = F) -->


<!-- summary(dmps.sign.anno.df) -->
<!-- colnames(dmps.sign.anno.df) -->
<!-- ``` -->

<!-- ```{r split-dmps-genes, include = F, eval = F} -->
<!-- library(splitstackshape) -->
<!-- tmp.df <- dmps.sign.anno.df[, .(PROBE_ID, UCSC_RefGene_Name)] -->

<!-- splitted.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Name', sep = ';', type.convert = F, drop = F) %>% -->
<!--   melt(id.vars = c("PROBE_ID", "UCSC_RefGene_Name"), value.name = "GeneSymbol") %>% -->
<!--   dplyr::select(PROBE_ID, GeneSymbol) %>% -->
<!--   unique() %>% -->
<!--   na.omit() -->

<!-- # Get coordintaes for genes -->

<!-- library(biomaRt) -->
<!-- ensembl      <- useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl')  -->
<!-- gene.anno.tbl <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'), -->
<!--                      filters = 'hgnc_symbol',  -->
<!--                      values = splitted.df$GeneSymbol,  -->
<!--                      mart = ensembl) %>%  -->
<!--   unique() -->

<!-- splitted.df <- left_join(splitted.df, gene.anno.tbl, by = c("GeneSymbol" = "hgnc_symbol")) -->
<!-- ``` -->

```{r load-sign-dmps, include = F, eval = T}
dmps.sign.anno.fn <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_fdr01_fc02_anno_full.csv"
dmps.sign.anno.df <- fread(dmps.sign.anno.fn)
dmps.sign.anno.df
```

## Overlap 450K and EPIC

```{r overlap-450-epic, out.width = "85%", fig.cap = "Overlap between 450K and EPIC significant DMPs ", include = T, eval = T}
library(eulerr)

dmps.450k.ids <- dmps.450k$Probe_Id
dmps.epic.ids <- dmps.sign.anno.df$PROBE_ID

fit <- euler(list(k450 = dmps.450k.ids, epic = dmps.epic.ids))

plot(fit,
     quantities = list(type = c("counts", "percent")),
     labels = c("450K", "EPIC"),
     fill = c("red", "pink", "transparent"),
     )
```
