***
# DMPs
***

## Roadmap

## Files location

## Signififcant DMPs

```{sh, include = F, eval = F}
scp -C -r -o ProxyCommand="ssh cluser@biocomp.psych.mpg.de nc slurmgate 22" ahryhorzhevska@slurmgate:/binder/mgp/datasets/2020_DexStim_Array_Human/methylation/20_DMA/01_lme_models/dnam_lmem_svs_pcs_rslt.txt output/data/methylation/01_lmem_dnam
```

```{r, include = F, eval = T}
source("~/bio/code/mpip/dex-stim-human-array/code/methylation/02_dma/01_dmp/funcitons.R")
```

```{r load-lmem-anno, include = F, eval = T}
lmem.rslt.fn <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/01_lmem_dnam/dnam_lmem_svs_pcs_rslt.txt'
anno.fn      <- '~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_cpgs_annotated.csv'

lmem.rslt    <- fread(lmem.rslt.fn)
anno.df      <- fread(anno.fn)
colnames(anno.df)[4] <- "PROBE_ID"
```


```{r load-pheno, include = F, eval = T}
pheno.full.fn <- paste0("~/bio/code/mpip/dex-stim-human-array/data/pheno/pheno_full_for_kimono.csv") 
pheno.full    <- read.csv2(pheno.full.fn) %>% setDT()

pheno <- pheno.full[Include == 1]

pheno$Status  <- as.factor(pheno$Status)
pheno$Sex     <- as.factor(pheno$Sex)

levels(pheno$Sex) <- c("Male", "Female")
levels(pheno$Status) <- c("Control", "MDD")
```

```{r load-dmps-450k, include = F, eval = T}
dmps.450k.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_q_value01FC002_methylation_450K_CellCode.txt"
dmps.450k    <- read.table(dmps.450k.fn, header = T)
```

```{r dmps-450k-inside, eval = T, include = F}
dmps.450k
```

```{r set-up-params, out.width = "85%", fig.cap = "Model with DNAm BCCs and Genotype PCs", include = T, eval = T}
fdr <- 0.01 #0.05
fc  <- 0.02 #0.01

VolcanoPlot(lmem.rslt, fdr, fc, "DNAm SVs and SNP PCs")
```

<!-- # Get significant dmps -->

<!-- ```{r save-dmps, include = F, eval = F} -->
<!-- dmps.sign.df <- lmem.rslt[abs(FC) > fc & FDR < fdr,] -->

<!-- dmps.sign.anno.df <- anno.df[PROBE_ID %in% dmps.sign.df$PROBE_ID ] -->
<!-- dmps.sign.anno.df <- left_join(dmps.sign.df, dmps.sign.anno.df, by = c("PROBE_ID" = "Name")) -->

<!-- fwrite(dmps.sign.anno.df,  -->
<!--        "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_fdr01_fc02_anno_full.csv",  -->
<!--        quote = F, row.names = F) -->


<!-- summary(dmps.sign.anno.df) -->
<!-- colnames(dmps.sign.anno.df) -->
<!-- ``` -->

<!-- ```{r split-dmps-genes, include = F, eval = F} -->
<!-- library(splitstackshape) -->
<!-- tmp.df <- dmps.sign.anno.df[, .(PROBE_ID, UCSC_RefGene_Name)] -->

<!-- splitted.df <- cSplit(indt = tmp.df, splitCols = 'UCSC_RefGene_Name', sep = ';', type.convert = F, drop = F) %>% -->
<!--   melt(id.vars = c("PROBE_ID", "UCSC_RefGene_Name"), value.name = "GeneSymbol") %>% -->
<!--   dplyr::select(PROBE_ID, GeneSymbol) %>% -->
<!--   unique() %>% -->
<!--   na.omit() -->

<!-- # Get coordintaes for genes -->

<!-- library(biomaRt) -->
<!-- ensembl      <- useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl')  -->
<!-- gene.anno.tbl <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'), -->
<!--                      filters = 'hgnc_symbol',  -->
<!--                      values = splitted.df$GeneSymbol,  -->
<!--                      mart = ensembl) %>%  -->
<!--   unique() -->

<!-- splitted.df <- left_join(splitted.df, gene.anno.tbl, by = c("GeneSymbol" = "hgnc_symbol")) -->
<!-- ``` -->

```{r load-sign-dmps, include = F, eval = T}
dmps.sign.anno.fn <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dmps_fdr01_fc02_anno_full.csv"
dmps.sign.anno.df <- fread(dmps.sign.anno.fn)
# dmps.sign.anno.df
```

## PCA

Relationship of the top 10 principal components derived from DNA methylation data with available and derived covariates. 

Because it is likely that other unmeasured environmental exposures also confound the analysis of methylomic variation associated with dex treatment, we investigated the impact of additional surrogate variables capturing variation in DNA methylation on the association statistics for dex-associated DMPs. We first compared each of the first 10 principal components (PCs) derived from the DNA methylation data to the available phenotype data to identify potential sources of additional variation between samples. For example, we observed strong correlations (r > 0.5) between the first three PCs and estimated blood cell composition measures, reflecting the likely effect of epigenetic differences between cell types. 

Heat-map of correlations between each PC (1-10) and the available phenotype information (SCZ – schizophrenia
status; sex) and variables derived from the DNA methylation data (DNAmAge – age estimated from DNA methylation data; DNAm Smoking - smoking score estimated from DNA methylation data; CD8T, CD4T, NK, Bcell, Mono, Gran - cellular proportion estimates from Houseman algorithm

```{r load-beta-mtrx, include = F, eval = F}
beta.mtrx.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_beta_combat_mtrx.rds"
beta.mtrx    <- readRDS(beta.mtrx.fn)
```

```{r load-nval-mtrx, include = F, eval = F}
mval.mtrx.fn <- "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_mval_combat_mtrx.rds"
mval.mtrx    <- readRDS(mval.mtrx.fn)

# Compute mval mtrx and save
# beta.mtrx   <- beta.mtrx[, match(pheno$DNAm_ID, colnames(beta.mtrx))]
# mval.mtrx   <- beta2m(beta.mtrx)
# saveRDS(mval.mtrx,
#        "~/bio/code/mpip/dex-stim-human-array/data/methylation/dex_methyl_mval_combat_mtrx.rds")
```

```{r compute-dnam-pcs, include = F, eval = F}
library(lumi)

t.mval.mtrx <- t(mval.mtrx[dmps.sign.anno.df$PROBE_ID,])
pca.obj <- prcomp(t.mval.mtrx, retx = T, center = T, scale. = F, rank. = 10)  
# pca.obj.retx <- prcomp(mval.mtrx, retx = T, center = T, scale. = F, rank. = 10)  

saveRDS(pca.obj,
         "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_methyl_pca_sign_dmps.rds")
```

```{r load-dnam-pcs, include = F, eval = T}
pca.obj.fn <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/02_dmp/dex_methyl_pca_sign_dmps.rds"
pca.obj    <- readRDS(pca.obj.fn)

nr.pcs   <- ncol(pca.obj$rotation)

propvar  <- round(summary(pca.obj)$importance["Proportion of Variance", 1:nr.pcs] * 100, 2 )
cummvar  <- round(summary(pca.obj)$importance["Cumulative Proportion", 1:nr.pcs] * 100, 2) 

dnam.pcs <- data.frame(pca.obj$x)
dnam.pcs <- dnam.pcs[match(pheno$DNAm_ID, as.factor(rownames(dnam.pcs))), ]

dnam.pcs    <- dnam.pcs[match(pheno$DNAm_ID, rownames(dnam.pcs)), paste0("PC", 1:10)] 
colnames(dnam.pcs) <- paste0("DNAm_PC", 1:10)
```

```{r combine-all-data, include = F, eval = T}
bcc.gex  <- pheno[, c("Cort_D1", "ACTH_D1", "Leukos_D1", "Gran_D1", "Mono_D1", "Lymph_D1")]
bcc.gex <- bcc.gex %>% 
  mutate_all(~ifelse(is.na(.), median(., na.rm = TRUE), .))
bcc.meth <- pheno[, c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Gran")]
sv.gex   <- pheno[, paste0("V", 1:10)] # c("V1", "V2", "V3", "V4", "V5")]
sv.meth  <- pheno[, paste0("DNAm_SV", 1:10)]# c("DNAm_SV1", "DNAm_SV2", "DNAm_SV3")]
snp.pcs  <- pheno[, paste0("PC", 1:5)]

tmp.df <- data.frame(sv.meth, meth = bcc.meth, gex = bcc.gex, gex = sv.gex, snp.pcs, dnam.pcs,
                     Dex = as.numeric(pheno$Dex), Status = as.numeric(pheno$Status), Sex = as.numeric(pheno$Sex), BMI = pheno$BMI_D1, 
                     Age = pheno$Age, DNAm_Age = pheno$mAge_Hovath, DNAm_Smoke = pheno$DNAm_SmokingScore) %>% 
  na.omit() %>% 
  scale()

cor.mtrx <- cor(tmp.df)
```

## Analysis {.tabset .tabset-pills}

### Corplot

```{r out.width = "98%", fig.cap = "Relationship of the top 10 PCs derived from DNAm data with with available and derived covariates", include = T, eval = T}
cor.mtrx.plt <- cor.mtrx[paste0("DNAm_PC", 1:10), c(paste0("DNAm_SV", 1:3),
                                                    paste("meth", colnames(bcc.meth), sep = "." ),
                                                    c("Dex", "Status", "Sex", "BMI", "Age", "DNAm_Age", "DNAm_Smoke"))]
par(cex = 1)
corrplot(cor.mtrx.plt, 
         method = "color", cl.pos = "b", cl.cex = 0.6,
         addgrid.col = "gray90",
         addCoef.col = "black", number.cex = 0.6, number.digits = 1, # Add coefficient of correlation
         tl.col = "black",  tl.cex = 0.6, tl.srt = 45, tl.offset = 0.5, tl.pos = "dt", #Text label color and rotation
         sig.level = 0.01) 
```

### Scree plot

```{r out.width = "90%", fig.cap = "Percentage of variance explained by first 10 eigenvalues"}
propvar.df <- data.frame(PC = 1:10, propvar)
ggplot(propvar.df, aes(PC, propvar)) +
  geom_bar(stat = "identity", width = 0.7, fill = "#00688B") +
  geom_line(colour = "red", linetype = "dashed") +
  geom_point(colour = "red") +
  xlab("DNAm PCs") +
  ylab("Percentage variance explained") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

### Dex

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the DEX treatment separation based on significant DMPs", include = T, eval = T}
plt.df <- data.frame((dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df, aes(DNAm_PC1, DNAm_PC2, col = Group)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("DNAm PC1 (", signif(propvar[1], 2), "%)")) +
  ylab(paste0("DNAm PC2 (", signif(propvar[2], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

### Sex

```{r out.width = "90%", fig.cap = "PCA plot demonstrating the SEX separation based on significant DMPs", include = T, eval = T}
plt.df <- data.frame((dnam.pcs), Group = as.factor(pheno$Dex), Sex = as.factor(pheno$Sex))

ggplot(data = plt.df[plt.df$Group == 1,], aes(DNAm_PC4, DNAm_PC7, col = Sex)) +
  geom_vline(xintercept = 0, linetype = "dashed", size = 0.4) +
  geom_hline(yintercept = 0, linetype = "dashed", size = 0.4) +
  geom_point(size = 2, alpha = 1) +
  xlab(paste0("DNAm PC4 (", signif(propvar[6], 2), "%)")) +
  ylab(paste0("DNAM PC7 (", signif(propvar[7], 2), "%)")) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1),
         legend.position = "none")
```

## Functional Annotation {.tabset .tabset-fade .tabset-pills}

```{r filter-data, include = F, eval = T}
dmps.sign.anno.df [dmps.sign.anno.df$FC < fc, MethylStatus := "hypomethylated"]
dmps.sign.anno.df [dmps.sign.anno.df$FC > fc, MethylStatus := "hypermethylated"]

interest.cols <- c("PROBE_ID", "chr", "pos", "UCSC_RefGene_Group", "UCSC_RefGene_Name", "Regulatory_Feature_Group", "Relation_to_Island",
                   "MethylStatus")

dmps.sign.sub.df       <- dmps.sign.anno.df %>% dplyr::select(interest.cols) %>% setDT()
dmps.sign.sub.df$Model <- "DMA"

dmps.merged.df       <- anno.df  %>% dplyr::select(interest.cols[-8]) %>% setDT()
dmps.merged.df$Model <- "Assessed"
dmps.merged.df       <- rbind(dmps.merged.df, dmps.sign.sub.df %>% dplyr::select(-MethylStatus))

# Split the UCSC RefGene Group

dmps.sign.sub.df$UCSC_RefGene_Group <- gsub(";.*", "", dmps.sign.sub.df$UCSC_RefGene_Group)
dmps.sign.sub.df[UCSC_RefGene_Group == "", UCSC_RefGene_Group := NA]

dmps.merged.df$UCSC_RefGene_Group <- gsub(";.*", "", dmps.merged.df$UCSC_RefGene_Group)
dmps.merged.df[UCSC_RefGene_Group == "", UCSC_RefGene_Group := NA]

```

### Chromosomes

```{r out.width = "90%", fig.cap = "DMPs distribution across chromosomes", include = T, eval = T}
chr.order <- paste("chr", c(1:22, "X", "Y"), sep = "")
meth      <- dmps.sign.sub.df
meth$chr  <- factor(meth$chr,levels = chr.order)

ggplot(meth, aes(x = chr)) + 
  geom_bar(aes(y = (..count..)/sum(..count..), fill = factor(..x..))) + 
  geom_text(aes(label = scales::percent((..count..)/sum(..count..), accuracy = 0.1), y = (..count..)/sum(..count..)), 
            stat = "count", vjust = -.5, size = 3) +
  scale_y_continuous(labels=scales::percent) +
  labs(x = "Chromosome",
       y = "Relative frequency") +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 45, hjust = 0.5),
         legend.position = "none")
```

### Genomic regions

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different genomic regions", include = T, eval = T}
ggplot(dmps.sign.sub.df, aes(x = Relation_to_Island, group = MethylStatus )) + 
  geom_bar(aes(y = ..prop.., fill = MethylStatus), stat = "count", position = position_dodge()) +
  geom_text(aes(label = scales::percent(..prop.., accuracy = 0.1), y = ..prop.. ), 
            stat = "count", position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Island",
       y = "Frequency", 
       title = "The ratio of hypermethylated versus hypomethylated dDMPs in different genomic regions") +
  theme(legend.title = element_blank(), 
        legend.position = c(.9,.85),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
  scale_fill_brewer(palette = "Dark1")
  # scale_fill_manual(values = c("grey", "darkgrey"))
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different genomic regions", incldue = F, eval = T}
ggplot(dmps.merged.df, aes(x = Relation_to_Island, group = Model )) + 
  geom_bar(aes(y = ..prop.., fill = Model), stat = "count", position = position_dodge()) +
  geom_text(aes(label = scales::percent(..prop.., accuracy = 0.1), y = ..prop.. ), 
            stat = "count", position = position_dodge(1.1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Island",
       y = "Frequency", 
       title = "The ration of all assessed EPIC CpGs versus dDMPs in different genomic regions") +
  theme(legend.title = element_blank(), 
        legend.position = c(.9,.75),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
  scale_fill_brewer(palette = "Set3")
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different genomic regions", include = T, eval = F}
grid.arrange(p1, p2, ncol = 2)
```

### Gene regions

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = T}
ggplot(dmps.sign.sub.df, aes(x = UCSC_RefGene_Group, group = MethylStatus )) + 
  geom_bar(aes(y = ..prop.., fill = MethylStatus), stat = "count", position = position_dodge()) +
  geom_text(aes(label = scales::percent(..prop.., accuracy = 0.1), y = ..prop.. ), 
            stat = "count", position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Island",
       y = "Frequency", 
       title = "The ratio of hypermethylated versus hypomethylated dDMPs in different genomic regions") +
  theme(legend.title = element_blank(), 
        legend.position = c(.9,.85),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
  scale_fill_brewer(palette = "Dark1")
  # scale_fill_manual(values = c("grey", "darkgrey"))
```

```{r out.width = "90%", fig.cap = "The distribution of CpGs in different gene regions", include = T, eval = T}
ggplot(dmps.merged.df, aes(x = UCSC_RefGene_Group, group = Model )) + 
  geom_bar(aes(y = ..prop.., fill = Model), stat = "count", position = position_dodge()) +
  geom_text(aes(label = scales::percent(..prop.., accuracy = 0.1), y = ..prop.. ), 
            stat = "count", position = position_dodge(1), size = 3) +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "Relation to Island",
       y = "Frequency", 
       title = "The ratio of all assessed EPIC CpGs versus dDMPs in different gene regions") +
  theme(legend.title = element_blank(), 
        legend.position = c(.9,.75),
        panel.grid.major = element_blank(),
        panel.background = element_blank(),
        plot.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 0.5)) +
  scale_fill_brewer(palette = "Set3")
```


## Overlap 450K and EPIC

```{r overlap-450-epic, out.width = "85%", fig.cap = "Overlap between 450K and EPIC significant DMPs ", include = T, eval = T}
library(eulerr)

dmps.450k.ids <- dmps.450k$Probe_Id
dmps.epic.ids <- dmps.sign.anno.df$PROBE_ID

fit <- euler(list(k450 = dmps.450k.ids, epic = dmps.epic.ids))

plot(fit,
     quantities = list(type = c("counts", "percent")),
     labels = c("450K", "EPIC"),
     fill = c("red", "pink", "transparent"),
     )
```

## GSEA

The GSEA was done using the rGREAT R package.

GREAT calculates statistics by associating genomic regions with nearby genes and applying the gene annotations to the regions. Association is a two step process. First, every gene is assigned a regulatory domain. Then, each genomic region is associated with all genes whose regulatory domain it overlaps.

Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS (regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene's basal domain but no more than the maximum extension in one direction.

### GO Enrichment for DMPs

Target genomic regions: significant DMPs
Background: all CpGs
Proximal: 5 kb upstream, 5 kb downstream, plus Distal: up to 10 kb

Statistical Significance: FDR <= 0.05

The test set contains 7,469 (1%) of all 740,357 regions.

The foreground set picked 1,613 genes, the background set picked 17,441 genes.

Ensembl Genes has 18,549 terms covering 18,549 (100%) of all 18,549 genes, and 18,549 term - gene associations.

18,549 ontology terms (100%) were tested using an annotation count range of [1, 1000].

```{r prepare-data-for-rgreat, icnldue = F, eval = F}
source("~/bio/code/mpip/dex-stim-human-array/code/methylation/02_dma/03_gsea/01_prepare_input_for_rgreat.R")

target.bed.fn <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/cpgs_target_for_great.bed"
bg.bed.fn     <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/cpgs_bg_for_great.bed"

# Export target data which are DMPs
bed.target <- ToBed(dmps.sign.anno.df, target.bed.fn, is.save = T)

# Export target data which are DMPs
bed.bg <- ToBed(anno.df, bg.bed.fn, is.save = T)
```

```{r submit-GREAT-job, include = F, eval = F}
library(rGREAT)

job <- submitGreatJob(bed.target, bg = bed.bg,
                      species = "hg19", rule = "basalPlusExt", adv_upstream = 5.0, adv_downstream = 5.0, adv_span = 10)

# availableOntologies(job)
great.enrichment.tbl <- getEnrichmentTables(job, download_by = "tsv",  ontology = c("GO Molecular Function", "GO Cellular Component", "GO Biological Process",
                                                                                    "Human Phenotype", "Ensembl Genes"))
great.cpg.gene.ass <- plotRegionGeneAssociationGraphs(job, plot = F)

saveRDS(object = job, 
        file = "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_job_svs_fdr01_fc02_output.rds")
saveRDS(object = great.enrichment.tbl, 
        file = "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_enrichmentTbl_svs_fdr01_fc02_output.rds")
saveRDS(object = great.cpg.gene.ass, 
        file = "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_cgGeneAssociationGRange_svs_fdr01_fc02_output.rds")
```

```{r load-rgreat-rslt, include = F, eval = T}
target.bed.fn <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/cpgs_target_for_great.bed"
bg.bed.fn     <- "~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/cpgs_bg_for_great.bed"

bed.target <- read.table(target.bed.fn, sep = " ")

job <- readRDS("~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_job_svs_fdr01_fc02_output.rds")
great.enrichment.tbl <- readRDS("~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_enrichmentTbl_svs_fdr01_fc02_output.rds")
great.cpg.gene.ass   <- readRDS("~/bio/code/mpip/dex-stim-human-array/output/data/methylation/04_gsea/rGREAT_rslt/rgreat_cgGeneAssociationGRange_svs_fdr01_fc02_output.rds")
```

```{r}
job
```

```{r}
plotRegionGeneAssociationGraphs(job, type = 1)
```
```{r}
plotRegionGeneAssociationGraphs(job, type = 2)
```
```{r}
plotRegionGeneAssociationGraphs(job, type = 3)
```
# Explore CpG - Gene associations

```{r}
cpg.gene.df <- data.frame(great.cpg.gene.ass) %>% select(chr = seqnames, start, end, gene, distTSS) %>% setDT() 
# unique(cpg.gene.df$gene) %>% length()

colnames(bed.target) <- c("chr", "start", "end", "name")

cpg.gene.df <- left_join(cpg.gene.df, bed.target)
# cpg.gene.df[!is.na(gene)][order(abs(distTSS))]

cpg.gene.df <- left_join(cpg.gene.df, dmps.sign.anno.df[, c("PROBE_ID", "FC", "MAD", "FDR")], by = c("name" = "PROBE_ID"))
datatable(cpg.gene.df[!is.na(gene)][order(abs(FDR))] %>% 
            mutate(absDistTSS = abs(distTSS)) %>%
            select(gene, cg = name, chr, start, distTSS, absDistTSS, FC, MAD, FDR))

```

# Explore GO Molecular Functions

```{r}
great.go.mol.fun <- great.enrichment.tbl[["GO Molecular Function"]] %>% setDT()

great.go.mol.fun[HyperFdrQ < p.thr][order(HyperFdrQ)] %>% select(ID, Desc, HyperFdrQ, RegionFoldEnrich, FgRegionsHit, NumFgGenesHit, FgGeneNames, BgRegionsHit, NumBgGenesHit)
```

```{r}
plot.df <- great.go.mol.fun[HyperFdrQ < p.thr] #[order(HyperFdrQ)] %>% select(HyperFdrQ, Desc) %>% mutate(HyperFdrQ = -log10(HyperFdrQ))

minus_log10_trans <- function(){
  trans_new(name = "minus_log10", transform = function(x) -log10(x), inverse = function(x) 10 ^ x)
}

plot.mf <- plot.df[1:20,]
plot.mf[["GO"]] <- "MF"  

ggplot(data = plot.df, aes(x = HyperFdrQ, y = reorder(Desc, -HyperFdrQ))) + 
  geom_bar(stat = "identity") +
  scale_x_continuous(trans = "log10",
                     labels = waiver(), limits = NULL) +# scales::number_format(accuracy = 0.000000000001)) +
  scale_y_discrete(position = "right") +
  labs(title = paste0("GO Molecular Function, FDR < ", p.thr),
       y = " ") # + geom_text(label = plot.df$HyperFdrQ)

```

# Explore GO Molecular Functions

```{r}
great.go.bio.proc <- great.enrichment.tbl[["GO Biological Process"]] %>% setDT()

(great.go.bio.proc[HyperFdrQ < p.thr][order(HyperFdrQ)] %>% dplyr::select(ID, Desc, HyperFdrQ, RegionFoldEnrich, FgRegionsHit, NumFgGenesHit, FgGeneNames, BgRegionsHit, NumBgGenesHit))
```
                                                                  

```{r}
plot.df <- great.go.bio.proc[HyperFdrQ < p.thr] #[order(HyperFdrQ)] %>% select(HyperFdrQ, Desc) %>% mutate(HyperFdrQ = -log10(HyperFdrQ))

plot.bp <- plot.df[1:20,]
plot.bp[["GO"]] <- "BP" 
  
ggplot(data = plot.df[1:20,], aes(x = HyperFdrQ, y = reorder(Desc, -HyperFdrQ))) + 
  geom_bar(stat = "identity") +
  scale_x_continuous(trans = "log10",
                     labels = waiver(), limits = NULL) +# scales::number_format(accuracy = 0.000000000001)) +
  scale_y_discrete(position = "right") +
  labs(title = paste0("Top 20 GO Biological Process with FDR < ", p.thr),
       y = " ") + # + geom_text(label = plot.df$HyperFdrQ) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1), 
         legend.position = "none")
```

```{r}
plot.df <- rbind(plot.bp, plot.mf)
plot.df$GO <- as.factor(plot.df$GO)
plot.df <- na.omit(plot.df)

ggplot(data = plot.df, aes(x = HyperFdrQ, y = reorder(Desc, -HyperFdrQ), fill = GO)) + 
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_x_continuous(trans = "log10",
                     labels = waiver(), limits = NULL) +# scales::number_format(accuracy = 0.000000000001)) +
  scale_y_discrete(position = "right") +
  labs(title = paste0("Top 40 GO with FDR < ", p.thr),
       y = " ") + # + geom_text(label = plot.df$HyperFdrQ) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1), 
         legend.position = c(.2, .2),
         legend.title = element_blank()) +
  scale_fill_brewer(palette = "Accent")
```

# Explore GO Cellular Component

```{r include=F}
great.go.cc <- great.enrichment.tbl[["GO Cellular Component"]] %>% setDT()

(great.go.cc[HyperFdrQ < p.thr][order(HyperFdrQ)] %>% dplyr::select(ID, Desc, HyperFdrQ, RegionFoldEnrich, FgRegionsHit, NumFgGenesHit, FgGeneNames, BgRegionsHit, NumBgGenesHit))
```

```{r include=T}
plot.df <- great.go.cc[HyperFdrQ < p.thr] #[order(HyperFdrQ)] %>% select(HyperFdrQ, Desc) %>% mutate(HyperFdrQ = -log10(HyperFdrQ))

plot.bp <- plot.df[1:20,]
plot.bp[["GO"]] <- "BP" 
  
ggplot(data = plot.df[1:20,], aes(x = HyperFdrQ, y = reorder(Desc, -HyperFdrQ))) + 
  geom_bar(stat = "identity") +
  scale_x_continuous(trans = "log10",
                     labels = waiver(), limits = NULL) +# scales::number_format(accuracy = 0.000000000001)) +
  scale_y_discrete(position = "right") +
  labs(title = paste0("Top 20 GO Biological Process with FDR < ", p.thr),
       y = " ") + # + geom_text(label = plot.df$HyperFdrQ) +
  theme( panel.background = element_blank(),
         plot.title = element_text(size = 10),
         axis.title = element_text(size = 10),
         axis.text.x = element_text(angle = 0, hjust = 1), 
         legend.position = "none")
```

# Explore ENSG

```{r}
p.thr <- 0.05

ensg.tbl <- great.enrichment.tbl[["Ensembl Genes"]] %>% setDT()
colnames(ensg.tbl)
ensg.tbl[HyperFdrQ <= p.thr][order(HyperFdrQ)]

cpg.gene.df.2 <- left_join(cpg.gene.df, ensg.tbl[, c("Desc", "HyperBonfP", "HyperFdrQ")], by = c("gene" = "Desc"))
datatable(cpg.gene.df.2[HyperFdrQ <= p.thr][order(HyperFdrQ)][,FDR := round(FDR, 30)])
```

# KEGG pathway enrichemnt analysis of DMPs


```{r gsea-missmethyl-kegg-all}
library(missMethyl)

gst.kegg <- gometh(sig.cpg = dmps.sign.anno.df$Name, all.cpg = dmps.anno.df$Name, array.type = "EPIC", anno = anno.epic.sub, collection = "KEGG", sig.genes = T)

write.csv2(gst.kegg, "~/bio/datasets/methylation/20_DMA/02_dmp/missMethyl_output/gsea_kegg_all_missmethyl.csv", quote = F, row.names = F)
```

```{r}
library(missMethyl)
gst.kegg <- read.csv2("~/bio/datasets/methylation/20_DMA/02_dmp/missMethyl_output/gsea_kegg_all_missmethyl.csv") %>% setDT()

kegg.sign.cat <- gst.kegg[FDR < 0.05][order(FDR)]
topGSA(kegg.sign.cat)
```

```{r gsea-missmethyl-kegg-promoter}
library(missMethyl)

gst.kegg.promoter <- gometh(sig.cpg = dmps.sign.anno.df$Name, all.cpg = dmps.anno.df$Name, array.type = "EPIC", anno = anno.epic.sub, 
                            collection = "KEGG", genomic.features = c("TSS200","TSS1500","1stExon"), sig.genes = T)

write.csv2(gst.kegg, "~/bio/datasets/methylation/20_DMA/02_dmp/missMethyl_output/gsea_kegg_promoter_missmethyl.csv", quote = F, row.names = F)
```

```{r}
gst.kegg <- read.csv2("~/bio/datasets/methylation/20_DMA/02_dmp/missMethyl_output/gsea_kegg_promoter_missmethyl.csv") %>% setDT()

kegg.sign.promoter.cat <- gst.kegg.promoter[FDR < 0.05][order(FDR)]
topGSA(kegg.sign.promoter.cat)
```

# DMRs obtained by comb-p

Dex-associated differentially methylated regions identified using CombP. Listed are all multiple testing significant (Sidak corrected P < 0.05) differentially methylated regions. 


```{r}
dmr.sign.combp.fn <- "~/bio/datasets/methylation/20_DMA/03_dmr/comb-p/dex.anno.hg19.regions.filtered.bed"
dmr.sign.combp.df <- fread(dmr.sign.combp.fn)

nr.probes <- 4

dmr.combp.sub.df <- dmr.sign.combp.df[n_probes >= nr.probes][z_sidak_p <= 0.05][order(z_sidak_p)] %>%
  dplyr::select(chr = `#chrom`, start, end, DMP_PVAL = min_p, Nr_Probes = n_probes, DMR_PVAL = z_sidak_p, Gene = refGene_name, Distance = refGene_distance, Feature = refGene_feature, CpG_Island_Name = cpgIslandExt_name, CpG_Island_Dist = cpgIslandExt_distance, CpG_Island_Feature = cpgIslandExt_feature)

summary(dmr.combp.sub.df)
fwrite(dmr.combp.sub.df[abs(end-start) <= 1000][order(DMP_PVAL)],
           file = "~/bio/datasets/methylation/20_DMA/03_dmr/comb-p/dex_top_dist_1kbp.txt",
           row.names = F, quote = F, dec = ".", col.names = T)
datatable(dmr.combp.sub.df[abs(end-start) <= 1000][order(DMP_PVAL)])

```